From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koyorice <87787929+Koyorice@users.noreply.github.com>
Date: Thu, 2 Oct 2025 18:33:36 +0800
Subject: [PATCH] Add MultiPaper Command


diff --git a/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java b/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
index 90ed57a7fbcd0625b64084347460e9864216f610..a0755c2475246f63a064c8d65d9b7e5500a5a2c5 100644
--- a/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
+++ b/src/main/java/org/bukkit/craftbukkit/command/CraftCommandMap.java
@@ -9,8 +9,25 @@ public class CraftCommandMap extends SimpleCommandMap {
 
     public CraftCommandMap(Server server) {
         super(server, io.papermc.paper.command.brigadier.bukkit.BukkitBrigForwardingMap.INSTANCE);
+        setDefaultCommands(); // MultiPaper - Add MultiPaper Command
     }
 
+    // MultiPaper start - Add MultiPaper Command
+    @Override
+    public void clearCommands() {
+        super.clearCommands();
+        setDefaultCommands();
+    }
+
+    private void setDefaultCommands() {
+        register("multipaper", new puregero.multipaper.commands.MPMapCommand("mpmap"));
+        register("multipaper", new puregero.multipaper.commands.SListCommand("slist"));
+        register("multipaper", new puregero.multipaper.commands.ServersCommand("servers"));
+        register("multipaper", new puregero.multipaper.commands.MPDebugCommand("mpdebug"));
+        register("multipaper", new puregero.multipaper.commands.EntitiesMapCommand("entitiesmap"));
+    }
+    // MultiPaper end - Add MultiPaper Command
+
     public Map<String, Command> getKnownCommands() {
         return this.knownCommands;
     }
