From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koyorice <87787929+Koyorice@users.noreply.github.com>
Date: Thu, 2 Oct 2025 16:58:17 +0800
Subject: [PATCH] Sync world


diff --git a/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java b/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
index 3e72dd125e5e7d577f0c1187d40b17cc6c2fea5a..b9b8d06380398340c982845814bdf451640c1362 100644
--- a/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
+++ b/src/main/java/io/papermc/paper/world/migration/WorldFolderMigration.java
@@ -44,6 +44,7 @@ public final class WorldFolderMigration {
         final ResourceKey<LevelStem> stemKey,
         final ResourceKey<Level> dimensionKey
     ) throws IOException {
+        if (true) return; // MultiPaper - Sync world
         final WorldMigrationContext context = new WorldMigrationContext(rootAccess, registryAccess, worldName, stemKey, dimensionKey);
         final MigrationMode mode = classifyStartupMigration(context);
         if (mode != MigrationMode.NO_OP) {
@@ -63,6 +64,7 @@ public final class WorldFolderMigration {
         final ResourceKey<LevelStem> stemKey,
         final ResourceKey<Level> dimensionKey
     ) throws IOException {
+        if (true) return; // MultiPaper - Sync world
         LegacyCraftBukkitWorldMigration.migrateApiWorld(new WorldMigrationContext(rootAccess, registryAccess, worldName, stemKey, dimensionKey));
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 6c02d75d615be10044b93dbf0c0870609a926ae6..10b5e6720f1710600d48f1d7d7c69a4b73163d57 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -1485,6 +1485,7 @@ public final class CraftServer implements Server {
             return;
         }
         this.worlds.put(world.getName().toLowerCase(Locale.ROOT), world);
+        puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(new puregero.multipaper.externalserverprotocol.SubscribeToWorldPacket(world.getUID())); // MultiPaper - Sync world
     }
 
     @Override
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index afc0a63c5f08de80915406c5a1f67eeb3a8f71a2..b952cf88dad5b5c85262cc3387813046ef0bf7bd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -830,6 +830,7 @@ public class CraftWorld extends CraftRegionAccessor implements World {
 
         // Updates the clock for all players
         this.world.clockManager().setTotalTicks(this.world.dimensionType().defaultClock().get(), currentClockTime + event.getSkipAmount());
+        puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(new puregero.multipaper.externalserverprotocol.TimeUpdatePacket(this.world, true)); // MultiPaper - Sync world
     }
 
     // Paper start
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index f3fd610c520dfe4b854a0cca7df0f4e3e1ef0da2..350f3f24ae91f46b51f6c72377993ae4fb2bced8 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -2344,6 +2344,7 @@ public class CraftEventFactory {
                 value = rule.deserialize(event.getValue()).getOrThrow(); // should never throw value is checked in the event
             }
             level.getGameRules().set(rule, value, level);
+            puregero.multipaper.externalserverprotocol.GameRuleUpdatePacket.onGameRuleChange(event.getWorld().getUID(), event.getGameRule().getName(), event.getValue()); // MultiPaper - Sync world
             return new GameRuleSetResult<>(value, false);
         } else {
             return new GameRuleSetResult<>(level.getGameRules().get(rule), true);
