From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koyorice <87787929+Koyorice@users.noreply.github.com>
Date: Mon, 6 Oct 2025 00:57:10 +0800
Subject: [PATCH] Sync entity


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
index ba37f6fdd3aef1328af7862def0ba8fd6af671a3..d4edcb7de58ab2321175e92e68a6efe13e8ed578 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -101,7 +101,7 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     protected Entity entity;
     private final EntityType entityType;
     private EntityDamageEvent lastDamageEvent;
-    private final CraftPersistentDataContainer persistentDataContainer = new CraftPersistentDataContainer(CraftEntity.DATA_TYPE_REGISTRY);
+    private final CraftPersistentDataContainer persistentDataContainer = new puregero.multipaper.MultiPaperDirtyCraftPersistentDataContainer(this, CraftEntity.DATA_TYPE_REGISTRY); // MultiPaper - Sync entity - sync persistent data container
     // Paper start - Folia shedulers
     public final io.papermc.paper.threadedregions.EntityScheduler taskScheduler = new io.papermc.paper.threadedregions.EntityScheduler(this);
     private final io.papermc.paper.threadedregions.scheduler.FoliaEntityScheduler apiScheduler = new io.papermc.paper.threadedregions.scheduler.FoliaEntityScheduler(this);
diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index 350f3f24ae91f46b51f6c72377993ae4fb2bced8..254b4cdc5396933e5ffb42399e4fde0958df4fe2 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -780,7 +780,7 @@ public class CraftEventFactory {
 
                 List<Entity> entities = level.getEntities(entity, entity.getBoundingBox().inflate(radius, radius, radius));
                 for (Entity e : entities) {
-                    if (e instanceof net.minecraft.world.entity.ExperienceOrb loopItem) {
+                    if (e instanceof net.minecraft.world.entity.ExperienceOrb loopItem && !puregero.multipaper.MultiPaper.isChunkExternal(level.getChunkIfLoaded(e.blockPosition()))) { // MultiPaper - Sync entity - don't merge exp orbs on external servers
                         // Paper start
                         if (!loopItem.isRemoved() && xp.count == loopItem.count && (mergeUnconditionally || loopItem.getValue() < maxValue) && new com.destroystokyo.paper.event.entity.ExperienceOrbMergeEvent((org.bukkit.entity.ExperienceOrb) entity.getBukkitEntity(), (org.bukkit.entity.ExperienceOrb) loopItem.getBukkitEntity()).callEvent()) { // Paper - ExperienceOrbMergeEvent
                             long newTotal = (long)xp.getValue() + (long)loopItem.getValue();
