From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koyorice <87787929+Koyorice@users.noreply.github.com>
Date: Fri, 3 Oct 2025 00:07:45 +0800
Subject: [PATCH] Sync chunk


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
index 3bb9b58ee97464687e348e23b99159226415c267..7bfbb7c7e6da9e67e021fa58901c5ce9bd44155e 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/MoonriseRegionFileIO.java
@@ -766,6 +766,12 @@ public final class MoonriseRegionFileIO {
     public static Cancellable loadDataAsync(final ServerLevel world, final int chunkX, final int chunkZ,
                                             final RegionFileType type, final BiConsumer<CompoundTag, Throwable> onComplete,
                                             final boolean intendingToBlock, final Priority priority) {
+        // MultiPaper start - Sync chunk
+        // event based io
+        if (puregero.multipaper.config.MultiPaperConfiguration.get().optimizations.useEventBasedIo) {
+            return puregero.multipaper.MultiPaperIO.loadDataAsync(world, chunkX, chunkZ, type, onComplete, intendingToBlock, priority);
+        }
+        // MultiPaper end - Sync chunk
         final RegionDataController taskController = getControllerFor(world, type);
 
         final ImmediateCallbackCompletion callbackInfo = new ImmediateCallbackCompletion();
@@ -873,6 +879,12 @@ public final class MoonriseRegionFileIO {
         }, true, priority);
 
         try {
+            // MultiPaper start - Sync chunk
+            // don't block the main thread
+            if (!ret.isDone() && org.bukkit.Bukkit.isPrimaryThread()) {
+                world.getChunkSource().mainThreadProcessor.managedBlock(ret::isDone);
+            }
+            // MultiPaper end - Sync chunk
             return ret.join();
         } catch (final CompletionException ex) {
             throw new IOException(ex);
@@ -1243,6 +1255,13 @@ public final class MoonriseRegionFileIO {
                     this.scheduleWriteCompress();
                     return;
                 }
+                // MultiPaper start - Sync chunk
+            } else if (writeData.result == RegionDataController.WriteData.WriteResult.SYNC_WRITE) {
+                if (!this.tryCompleteWrite(inProgressWrite, failedWrite)) {
+                    // fetch new data and try again
+                    this.scheduleWriteCompress();
+                }
+                // MultiPaper end - Sync chunk
             } else {
                 // writeData != null && !failedWrite
                 // we can continue to I/O stage
@@ -1466,6 +1485,7 @@ public final class MoonriseRegionFileIO {
         public static record WriteData(CompoundTag input, WriteResult result, DataOutputStream output, IORunnable write) {
             public static enum WriteResult {
                 WRITE,
+                SYNC_WRITE, // MultiPaper - Sync chunk
                 DELETE;
             }
         }
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/ChunkDataController.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/ChunkDataController.java
index a3250959137c559ddfc78833b50f0aa51b2639a7..658a535559bc929078a67754d38ab9a9601cefff 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/ChunkDataController.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/ChunkDataController.java
@@ -25,21 +25,30 @@ public final class ChunkDataController extends MoonriseRegionFileIO.RegionDataCo
 
     @Override
     public WriteData startWrite(final int chunkX, final int chunkZ, final CompoundTag compound) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$startWrite(chunkX, chunkZ, compound);
+        // MultiPaper start - Sync chunk
+        puregero.multipaper.MultiPaper.writeRegionFileNBT(this.world.getWorld().getName(), this.getCache().folderStr, "region", chunkX, chunkZ, compound);
+        return new WriteData(null, WriteData.WriteResult.SYNC_WRITE, null, null);
+        // MultiPaper end - Sync chunk
     }
 
     @Override
     public void finishWrite(final int chunkX, final int chunkZ, final WriteData writeData) throws IOException {
-        ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$finishWrite(chunkX, chunkZ, writeData);
+        throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
     }
 
     @Override
     public ReadData readData(final int chunkX, final int chunkZ) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$readData(chunkX, chunkZ);
+        // MultiPaper start - Sync chunk
+        CompoundTag compound = puregero.multipaper.MultiPaper.readRegionFileNBT(this.world.getWorld().getName(), this.getCache().folderStr, "region", chunkX, chunkZ);
+        if (compound == null) {
+            return new ReadData(ReadData.ReadResult.NO_DATA, null, null, 0);
+        }
+        return new ReadData(ReadData.ReadResult.SYNC_READ, null, compound, 0);
+        // MultiPaper end - Sync chunk
     }
 
     @Override
     public CompoundTag finishRead(final int chunkX, final int chunkZ, final ReadData readData) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$finishRead(chunkX, chunkZ, readData);
+        throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
     }
 }
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/EntityDataController.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/EntityDataController.java
index 992d19e1b08fa84325e7bb3c3d7d0b9dff4cdd3b..5caeb517008e9dad5bb95bee6ca455371446e4d6 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/EntityDataController.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/EntityDataController.java
@@ -14,10 +14,12 @@ import java.nio.file.Path;
 public final class EntityDataController extends MoonriseRegionFileIO.RegionDataController {
 
     private final EntityRegionFileStorage storage;
+    private final String world; // MultiPaper - Sync chunk
 
-    public EntityDataController(final EntityRegionFileStorage storage, final ChunkTaskScheduler taskScheduler) {
+    public EntityDataController(final EntityRegionFileStorage storage, final ChunkTaskScheduler taskScheduler, final String world) { // MultiPaper - Sync chunk
         super(MoonriseRegionFileIO.RegionFileType.ENTITY_DATA, taskScheduler.ioExecutor, taskScheduler.compressionExecutor);
         this.storage = storage;
+        this.world = world; // MultiPaper - Sync chunk
     }
 
     @Override
@@ -29,22 +31,31 @@ public final class EntityDataController extends MoonriseRegionFileIO.RegionDataC
     public WriteData startWrite(final int chunkX, final int chunkZ, final CompoundTag compound) throws IOException {
         checkPosition(new ChunkPos(chunkX, chunkZ), compound);
 
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$startWrite(chunkX, chunkZ, compound);
+        // MultiPaper start - Sync chunk
+        puregero.multipaper.MultiPaper.writeRegionFileNBT(this.world, this.getCache().folderStr, "entities", chunkX, chunkZ, compound);
+        return new WriteData(null, WriteData.WriteResult.SYNC_WRITE, null, null);
+        // MultiPaper end - Sync chunk
     }
 
     @Override
     public void finishWrite(final int chunkX, final int chunkZ, final WriteData writeData) throws IOException {
-        ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$finishWrite(chunkX, chunkZ, writeData);
+        throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
     }
 
     @Override
     public ReadData readData(final int chunkX, final int chunkZ) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$readData(chunkX, chunkZ);
+        // MultiPaper start - Sync chunk
+        CompoundTag compound = puregero.multipaper.MultiPaper.readRegionFileNBT(this.world, this.getCache().folderStr, "entities", chunkX, chunkZ);
+        if (compound == null) {
+            return new ReadData(ReadData.ReadResult.NO_DATA, null, null, 0);
+        }
+        return new ReadData(ReadData.ReadResult.SYNC_READ, null, compound, 0);
+        // MultiPaper end - Sync chunk
     }
 
     @Override
     public CompoundTag finishRead(final int chunkX, final int chunkZ, final ReadData readData) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$finishRead(chunkX, chunkZ, readData);
+        throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
     }
 
     private static void checkPosition(final ChunkPos pos, final CompoundTag nbt) {
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/PoiDataController.java b/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/PoiDataController.java
index bd0d782852f9cfe5bc0b5339ecf4d82c10332ec9..1d37ff6ac199871ea586b18d63d4af6b0de68328 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/PoiDataController.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/io/datacontroller/PoiDataController.java
@@ -25,21 +25,30 @@ public final class PoiDataController extends MoonriseRegionFileIO.RegionDataCont
 
     @Override
     public WriteData startWrite(final int chunkX, final int chunkZ, final CompoundTag compound) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$startWrite(chunkX, chunkZ, compound);
+        // MultiPaper start - Sync chunk
+        puregero.multipaper.MultiPaper.writeRegionFileNBT(this.world.getWorld().getName(), this.getCache().folderStr, "poi", chunkX, chunkZ, compound);
+        return new WriteData(null, WriteData.WriteResult.SYNC_WRITE, null, null);
+        // MultiPaper end - Sync chunk
     }
 
     @Override
     public void finishWrite(final int chunkX, final int chunkZ, final WriteData writeData) throws IOException {
-        ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$finishWrite(chunkX, chunkZ, writeData);
+        throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
     }
 
     @Override
     public ReadData readData(final int chunkX, final int chunkZ) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$readData(chunkX, chunkZ);
+        // MultiPaper start - Sync chunk
+        CompoundTag compound = puregero.multipaper.MultiPaper.readRegionFileNBT(this.world.getWorld().getName(), this.getCache().folderStr, "poi", chunkX, chunkZ);
+        if (compound == null) {
+            return new ReadData(ReadData.ReadResult.NO_DATA, null, null, 0);
+        }
+        return new ReadData(ReadData.ReadResult.SYNC_READ, null, compound, 0);
+        // MultiPaper end - Sync chunk
     }
 
     @Override
     public CompoundTag finishRead(final int chunkX, final int chunkZ, final ReadData readData) throws IOException {
-        return ((ChunkSystemRegionFileStorage)this.getCache()).moonrise$finishRead(chunkX, chunkZ, readData);
+        throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
     }
 }
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
index e418820359c430b91bb91eb65d137b9cdbe2db4b..4c48c04997d3ec646543a449108411c42a4dbbcb 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/ChunkHolderManager.java
@@ -1105,6 +1105,12 @@ public final class ChunkHolderManager {
 
             // Note: no need to busy wait on the chunk queue, poi load will complete off-main
 
+            // MultiPaper start - Sync chunk
+            // don't block the main thread
+            if (!isCompleted.get() && org.bukkit.Bukkit.isPrimaryThread()) {
+                this.world.getChunkSource().mainThreadProcessor.managedBlock(isCompleted::get);
+            }
+            // MultiPaper end - Sync chunk
             boolean interrupted = false;
             while (!isCompleted.get()) {
                 interrupted |= Thread.interrupted();
@@ -1148,6 +1154,65 @@ public final class ChunkHolderManager {
         this.chunkHolders.remove(CoordinateUtils.getChunkKey(holder.chunkX, holder.chunkZ));
     }
 
+    // MultiPaper start - Sync chunk
+    public void unloadChunkNowNoSave(NewChunkHolder holder) {
+        TickThread.ensureTickThread("Cannot unload chunks off-main");
+
+        if (BLOCK_TICKET_UPDATES.get() == Boolean.TRUE) {
+            throw new IllegalStateException("Cannot unload chunks recursively");
+        }
+        if (this.ticketLockArea.isHeldByCurrentThread(holder.chunkX, holder.chunkZ)) {
+            throw new IllegalStateException("Cannot hold ticket update lock while calling processUnloads");
+        }
+        if (this.taskScheduler.schedulingLockArea.isHeldByCurrentThread(holder.chunkX, holder.chunkZ)) {
+            throw new IllegalStateException("Cannot hold scheduling lock while calling processUnloads");
+        }
+
+        NewChunkHolder.UnloadState state;
+
+        ReentrantAreaLock.Node ticketLock = this.ticketLockArea.lock(holder.chunkX, holder.chunkZ);
+        try {
+            final ReentrantAreaLock.Node scheduleLock = this.taskScheduler.schedulingLockArea.lock(holder.chunkX, holder.chunkZ);
+            try {
+                state = holder.unloadStage1();
+                if (state == null) {
+                    // can unload immediately
+                    this.removeChunkHolder(holder);
+                    return;
+                }
+            } finally {
+                this.taskScheduler.schedulingLockArea.unlock(scheduleLock);
+            }
+        } finally {
+            this.ticketLockArea.unlock(ticketLock);
+        }
+
+        final Boolean before = this.blockTicketUpdates();
+        try {
+            holder.unloadStage2(state, true);
+        } finally {
+            this.unblockTicketUpdates(before);
+        }
+
+        ticketLock = this.ticketLockArea.lock(holder.chunkX, holder.chunkZ);
+        try {
+            final ReentrantAreaLock.Node scheduleLock = this.taskScheduler.schedulingLockArea.lock(holder.chunkX, holder.chunkZ);
+            try {
+                if (holder.unloadStage3()) {
+                    this.removeChunkHolder(holder);
+                } else {
+                    // add cooldown so the next unload check is not immediately next tick
+                    this.addTicketAtLevel(UNLOAD_COOLDOWN, CoordinateUtils.getChunkKey(holder.chunkX, holder.chunkZ), MAX_TICKET_LEVEL, null, true);
+                }
+            } finally {
+                this.taskScheduler.schedulingLockArea.unlock(scheduleLock);
+            }
+        } finally {
+            this.ticketLockArea.unlock(ticketLock);
+        }
+    }
+    // MultiPaper end - Sync chunk
+
     // note: never call while inside the chunk system, this will absolutely break everything
     public void processUnloads() {
         TickThread.ensureTickThread("Cannot unload chunks off-main");
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
index 9a9a599ef178f851ee5c783631a724013a693586..203447556dda19df6a6ea812675253362cf12d45 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
@@ -68,6 +68,13 @@ public final class NewChunkHolder {
 
     public final ChunkTaskScheduler scheduler;
 
+    // MultiPaper start - Sync chunk
+    @javax.annotation.Nullable
+    public puregero.multipaper.ExternalServer externalOwner; // Chunk owner
+    public boolean hasExternalLockRequest = false; // Have we sent a request to lock this chunk?
+    public java.util.Set<puregero.multipaper.ExternalServer> externalSubscribers = java.util.concurrent.ConcurrentHashMap.newKeySet(); // Chunk subscribers
+    // MultiPaper end - Sync chunk
+
     // load/unload state
 
     // chunk data state
@@ -451,6 +458,25 @@ public final class NewChunkHolder {
 
     private ChunkAccess currentChunk;
 
+    // MultiPaper start - Sync chunk
+    private java.util.concurrent.CompletableFuture<ChunkAccess> onCurrentChunkLoaded;
+
+    public java.util.concurrent.CompletableFuture<ChunkAccess> onCurrentChunkLoaded() {
+        if (!this.hasGenerationTask()) {
+            if (onCurrentChunkLoaded == null || (onCurrentChunkLoaded.isDone() && onCurrentChunkLoaded.getNow(null) != currentChunk)) {
+                onCurrentChunkLoaded = java.util.concurrent.CompletableFuture.completedFuture(currentChunk);
+            }
+            return onCurrentChunkLoaded;
+        }
+
+        if (onCurrentChunkLoaded == null || onCurrentChunkLoaded.isDone()) {
+            onCurrentChunkLoaded = new java.util.concurrent.CompletableFuture<>();
+        }
+
+        return onCurrentChunkLoaded;
+    }
+    // MultiPaper end - Sync chunk
+
     // generation status state
 
     /**
@@ -867,12 +893,18 @@ public final class NewChunkHolder {
     }
 
     void unloadStage2(final UnloadState state) {
+        // MultiPaper start - Sync chunk
+        this.unloadStage2(state, false);
+    }
+
+    void unloadStage2(final UnloadState state, boolean doNotSaveChunk) {
+        // MultiPaper end - Sync chunk
         this.unloadState = null;
         final ChunkAccess chunk = state.chunk();
         final ChunkEntitySlices entityChunk = state.entityChunk();
         final PoiChunk poiChunk = state.poiChunk();
 
-        final boolean shouldLevelChunkNotSave = PlatformHooks.get().forceNoSave(chunk);
+        final boolean shouldLevelChunkNotSave = doNotSaveChunk || PlatformHooks.get().forceNoSave(chunk); // MultiPaper - Sync chunk - add an option to not save chunk
 
         // unload chunk data
         if (chunk != null) {
@@ -891,12 +923,13 @@ public final class NewChunkHolder {
             }
 
             this.asyncCompleteUnloadTask(MoonriseRegionFileIO.RegionFileType.CHUNK_DATA, chunkWrite[0]);
+            puregero.multipaper.MultiPaperChunkHandler.onChunkUnload(this, chunk, entityChunk); // MultiPaper - Sync chunk
         }
 
         // unload entity data
         if (entityChunk != null) {
             final Completable<CompoundTag>[] entityWrite = new Completable[1];
-            this.saveEntities(entityChunk, true, entityWrite);
+            if (!doNotSaveChunk) this.saveEntities(entityChunk, true, entityWrite); // MultiPaper - Sync chunk
 
             if (entityChunk.unload()) {
                 final ReentrantAreaLock.Node schedulingLock = this.scheduler.schedulingLockArea.lock(this.chunkX, this.chunkZ);
@@ -1169,6 +1202,7 @@ public final class NewChunkHolder {
                 for (int dx = -NEIGHBOUR_RADIUS; dx <= NEIGHBOUR_RADIUS; ++dx) {
                     final NewChunkHolder holder = (dx | dz) == 0 ? this : this.scheduler.chunkHolderManager.getChunkHolder(dx + this.chunkX, dz + this.chunkZ);
                     if (loaded) {
+                        if (holder == null) LOGGER.warn("Holder for ({},{}) is null while updating full status for ({},{})", dx + this.chunkX, dz + this.chunkZ, this.chunkX, this.chunkZ); // MultiPaper - Sync chunk
                         if (holder.setNeighbourFullLoaded(-dx, -dz)) {
                             changedFullStatus.add(holder);
                         }
@@ -1272,6 +1306,7 @@ public final class NewChunkHolder {
                     if (current.isOrAfter(FullChunkStatus.ENTITY_TICKING) && !pending.isOrAfter(FullChunkStatus.ENTITY_TICKING)) {
                         this.changeEntityChunkStatus(FullChunkStatus.BLOCK_TICKING);
                         PlatformHooks.get().onChunkNotEntityTicking(chunk, this.vanillaChunkHolder);
+                        if (this.hasExternalLockRequest) puregero.multipaper.MultiPaper.unlockChunk(this, this.getCurrentChunk(), this.getEntityChunk()); // MultiPaper - Sync chunk
                         this.updateCurrentState(FullChunkStatus.BLOCK_TICKING);
                     }
 
@@ -1464,6 +1499,7 @@ public final class NewChunkHolder {
         final ChunkCompletion completion = new ChunkCompletion(newChunk, newStatus);
         CHUNK_COMPLETION_ARRAY_HANDLE.setVolatile(this.chunkCompletions, newStatus.getIndex(), completion);
         this.lastChunkCompletion = completion;
+        puregero.multipaper.MultiPaper.getConnection().send(new puregero.multipaper.mastermessagingprotocol.messages.masterbound.SubscribeChunkMessage(world.getWorld().getName(), chunkX, chunkZ)); // MultiPaper - Sync chunk - ensure we're subscribed to this chunk
 
         final ChunkStatus requestedGenStatus = this.requestedGenStatus;
 
@@ -1542,6 +1578,8 @@ public final class NewChunkHolder {
 
             this.scheduleNeighbours(needsScheduling, scheduleList);
         }
+
+        if (onCurrentChunkLoaded != null) onCurrentChunkLoaded.completeAsync(() -> currentChunk, world.getServer()::scheduleOnMain); // MultiPaper - Sync chunk - listen for loading of the current chunk
     }
 
     private void scheduleNeighbours(final List<NewChunkHolder> needsScheduling, final List<ChunkProgressionTask> scheduleList) {
@@ -1739,6 +1777,7 @@ public final class NewChunkHolder {
 
             task.queue();
 
+            puregero.multipaper.MultiPaper.willSaveChunk(this.world, this.chunkX, this.chunkZ); // MultiPaper - Sync Chunk
             MoonriseRegionFileIO.scheduleSave(
                 this.world, this.chunkX, this.chunkZ, completable, task, MoonriseRegionFileIO.RegionFileType.CHUNK_DATA, Priority.NORMAL
             );
@@ -1766,9 +1805,7 @@ public final class NewChunkHolder {
                     // we have no entities to add to the existing chunk
                     return false;
                 }
-                final Completable<CompoundTag> mergeFrom = MoonriseRegionFileIO.loadDataAsync(
-                    this.world, this.chunkX, this.chunkZ, MoonriseRegionFileIO.RegionFileType.ENTITY_DATA, false, Priority.NORMAL
-                );
+                final Completable<CompoundTag> mergeFrom = Completable.completed(new CompoundTag()); // MultiPaper - Sync chunk - process transient entity chunks on the master
 
                 final Completable<CompoundTag> toWrite = mergeFrom.handle((final CompoundTag onDisk, final Throwable thr) -> {
                     if (thr != null) {
@@ -1776,6 +1813,7 @@ public final class NewChunkHolder {
                         return save;
                     } else {
                         ChunkEntitySlices.copyEntities(onDisk, save);
+                        save.putBoolean("multipaper.transient", true); // MultiPaper - Sync chunk - process transient entity chunks on the master
                         return save;
                     }
                 });
diff --git a/net/minecraft/network/protocol/game/ClientboundBlockEntityDataPacket.java b/net/minecraft/network/protocol/game/ClientboundBlockEntityDataPacket.java
index 2122ff56412a2ad412eb8b2f0e4538aab4257bf2..a84e33dec4d7c0a478b55363e1bdd7b3afa89c78 100644
--- a/net/minecraft/network/protocol/game/ClientboundBlockEntityDataPacket.java
+++ b/net/minecraft/network/protocol/game/ClientboundBlockEntityDataPacket.java
@@ -30,8 +30,15 @@ public class ClientboundBlockEntityDataPacket implements Packet<ClientGamePacket
     public static ClientboundBlockEntityDataPacket create(
         final BlockEntity blockEntity, final BiFunction<BlockEntity, RegistryAccess, CompoundTag> updateTagSaver
     ) {
+    // MultiPaper start - Sync chunk
+        return create(blockEntity, updateTagSaver, true);
+    }
+        public static ClientboundBlockEntityDataPacket create(
+        final BlockEntity blockEntity, final BiFunction<BlockEntity, RegistryAccess, CompoundTag> updateTagSaver, final boolean sanitizeNbt
+    ) {
+    // MultiPaper end - Sync chunk
         RegistryAccess registryAccess = blockEntity.getLevel().registryAccess();
-        return new ClientboundBlockEntityDataPacket(blockEntity.getBlockPos(), blockEntity.getType(), blockEntity.sanitizeSentNbt(updateTagSaver.apply(blockEntity, registryAccess))); // Paper - Sanitize sent data
+        return new ClientboundBlockEntityDataPacket(blockEntity.getBlockPos(), blockEntity.getType(), sanitizeNbt ? blockEntity.sanitizeSentNbt(updateTagSaver.apply(blockEntity, registryAccess)) : updateTagSaver.apply(blockEntity, registryAccess)); // Paper - Sanitize sent data // MultiPaper - Sync chunk - skip nbt sanitization when syncing between servers
     }
 
     public static ClientboundBlockEntityDataPacket create(final BlockEntity blockEntity) {
diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index d9be20855c952c49555ffc424318a7f1c32f1005..5e3094f9521bc9fba41bc8a512ce0412462375d3 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -759,6 +759,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
         overworld.getWorldBorder().world = overworld;
         overworld.getWorldBorder().setAbsoluteMaxSize(this.getAbsoluteMaxWorldSize());
+        overworld.getWorldBorder().addListener(new puregero.multipaper.MultiPaperWorldBorderHandler()); // MultiPaper - Sync chunk - add our own world border listener
         this.getPlayerList().addWorldborderListener(overworld);
         this.server.getPluginManager().callEvent(new org.bukkit.event.world.WorldInitEvent(overworld.getWorld()));
     // Paper end - rework world loading process
@@ -1080,6 +1081,16 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         }
         // CraftBukkit end
         this.getConnection().stop();
+        // MultiPaper start - Sync chunk
+        for (ServerLevel world : this.getAllLevels()) {
+            ((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemServerLevel)world).moonrise$getEntityTickingChunks().iterator().forEachRemaining(chunk -> {
+                ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder chunkHolder = chunk.moonrise$getChunkHolder();
+                if (chunkHolder.hasExternalLockRequest) {
+                    puregero.multipaper.MultiPaper.unlockChunk(chunkHolder, chunk, chunkHolder.getEntityChunk());
+                }
+            });
+        }
+        // MultiPaper end - Sync chunk
         this.isSaving = true;
         if (this.playerList != null) {
             LOGGER.info("Saving players");
@@ -1891,6 +1902,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         this.updateEffectiveRespawnData();
 
         this.isIteratingOverLevels = true; // Paper - Throw exception on world create while being ticked
+        puregero.multipaper.MultiPaperExternalBlocksHandler.tick(); // MultiPaper - Sync chunk
         for (ServerLevel level : this.getAllLevels()) {
             level.hasPhysicsEvent = org.bukkit.event.block.BlockPhysicsEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - BlockPhysicsEvent
             level.hasEntityMoveEvent = io.papermc.paper.event.entity.EntityMoveEvent.getHandlerList().getRegisteredListeners().length > 0; // Paper - Add EntityMoveEvent
diff --git a/net/minecraft/server/level/ChunkHolder.java b/net/minecraft/server/level/ChunkHolder.java
index cd5a325d37f617431ceb568a655701601618198f..f0f8c1730526f9d371fe4c3c6fbebe401474e39b 100644
--- a/net/minecraft/server/level/ChunkHolder.java
+++ b/net/minecraft/server/level/ChunkHolder.java
@@ -33,6 +33,12 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
     public static final ChunkResult<LevelChunk> UNLOADED_LEVEL_CHUNK = ChunkResult.error("Unloaded level chunk");
     private static final CompletableFuture<ChunkResult<LevelChunk>> UNLOADED_LEVEL_CHUNK_FUTURE = CompletableFuture.completedFuture(UNLOADED_LEVEL_CHUNK);
     private final LevelHeightAccessor levelHeightAccessor;
+    // MultiPaper start - Sync chunk
+    public boolean addChangesToPlayersOnly = false;
+    private boolean broadcastingToOtherServersOnly = false;
+    private boolean broadcastingToPlayersOnly = false;
+    private java.util.@Nullable Set<BlockPos> blocksToBroadcastToPlayers = null;
+    // MultiPaper end - Sync chunk
     // Paper - rewrite chunk system
     private boolean hasChangedSections;
     private final @Nullable ShortSet[] changedBlocksPerSection;
@@ -221,7 +227,14 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
     }
 
     public boolean blockChanged(final BlockPos pos) {
-        LevelChunk chunk = this.playersSentChunkTo.size() == 0 ? null : this.getChunkToSend(); // Paper - rewrite chunk system
+        // MultiPaper start - Sync chunk
+        if (this.addChangesToPlayersOnly && !puregero.multipaper.MultiPaper.isChunkLocal(this.getChunkMap().level.getChunkIfLoaded(pos))) {
+            this.broadcastBlockToPlayersLater(pos);
+            return false;
+        }
+        // if we modify this chunk, even if it's not sent to players, we still need to send it to other servers
+        LevelChunk chunk = this.getChunkMap().level.getChunkSource().getChunkAtIfLoadedImmediately(this.pos.x(), this.pos.z());
+        // MultiPaper end - Sync chunk
         if (chunk == null) {
             return false;
         }
@@ -237,6 +250,7 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
         }
 
         changedBlocksInSection.add(SectionPos.sectionRelativePos(pos));
+        if (this.blocksToBroadcastToPlayers != null && !this.blocksToBroadcastToPlayers.isEmpty()) this.blocksToBroadcastToPlayers.remove(pos); // MultiPaper - Sync chunk
         return !hadChangedSections;
     }
 
@@ -272,7 +286,45 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
         return this.hasChangedSections || !this.skyChangedLightSectionFilter.isEmpty() || !this.blockChangedLightSectionFilter.isEmpty();
     }
 
+    // MultiPaper start - Sync chunk
+    public void broadcastChangesToOtherServers(LevelChunk chunk) {
+        if (newChunkHolder.externalSubscribers.isEmpty()) return; // No point running this if no one's listening
+        broadcastingToOtherServersOnly = true;
+        broadcastChanges(chunk);
+        broadcastingToOtherServersOnly = false;
+    }
+
+    private void broadcastBlockToPlayersLater(BlockPos blockPos) {
+        if (blocksToBroadcastToPlayers == null) {
+            blocksToBroadcastToPlayers = new it.unimi.dsi.fastutil.objects.ObjectLinkedOpenHashSet<>();
+        }
+
+        if (blockPos instanceof BlockPos.MutableBlockPos) {
+            // Let's make it not mutable...
+            blockPos = new BlockPos(blockPos);
+        }
+
+        this.getChunkMap().level.getChunkSource().addChunkHolderUnsafe(this);
+        this.blocksToBroadcastToPlayers.add(blockPos);
+    }
+    // MultiPaper end - Sync chunk
+
     public void broadcastChanges(final LevelChunk chunk) {
+        // MultiPaper start - Sync chunk
+        if (!this.broadcastingToOtherServersOnly && this.blocksToBroadcastToPlayers != null && !this.blocksToBroadcastToPlayers.isEmpty()) {
+            // Send blocks to players that shouldn't be sent to other servers as they've already been sent to other servers
+            this.broadcastingToPlayersOnly = true;
+            this.blocksToBroadcastToPlayers.forEach(blockPos -> {
+                Level level = chunk.getLevel();
+                BlockState blockState = level.getBlockState(blockPos);
+                List<ServerPlayer> players = this.moonrise$getPlayers(false);
+                this.broadcast(players, new ClientboundBlockUpdatePacket(blockPos, blockState));
+                this.broadcastBlockEntityIfNeeded(players, level, blockPos, blockState);
+            });
+            this.broadcastingToPlayersOnly = false;
+            this.blocksToBroadcastToPlayers.clear();
+        }
+        // MultiPaper end - Sync chunk
         if (this.hasChangesToBroadcast()) {
             Level level = chunk.getLevel();
             if (!this.skyChangedLightSectionFilter.isEmpty() || !this.blockChangedLightSectionFilter.isEmpty()) {
@@ -288,26 +340,32 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
                 this.blockChangedLightSectionFilter.clear();
             }
 
-            if (this.hasChangedSections) {
+            if (!broadcastingToOtherServersOnly && this.hasChangedSections) { // MultiPaper - Sync chunk
                 List<ServerPlayer> players = this.moonrise$getPlayers(false); // Paper - rewrite chunk system
 
                 for (int sectionIndex = 0; sectionIndex < this.changedBlocksPerSection.length; sectionIndex++) {
                     ShortSet changedBlocks = this.changedBlocksPerSection[sectionIndex];
                     if (changedBlocks != null) {
                         this.changedBlocksPerSection[sectionIndex] = null;
-                        if (!players.isEmpty()) {
+                        if (true) { // MultiPaper - Sync chunk - send block changes even if there are no players, as there might be servers to send them to
                             int sectionY = this.levelHeightAccessor.getSectionYFromSectionIndex(sectionIndex);
                             SectionPos sectionPos = SectionPos.of(chunk.getPos(), sectionY);
                             if (changedBlocks.size() == 1) {
                                 BlockPos pos = sectionPos.relativeToBlockPos(changedBlocks.iterator().nextShort());
                                 BlockState state = level.getBlockState(pos);
+                                if (this.broadcastingToOtherServersOnly) broadcastBlockToPlayersLater(pos); // MultiPaper - Sync chunk
                                 this.broadcast(players, new ClientboundBlockUpdatePacket(pos, state));
                                 this.broadcastBlockEntityIfNeeded(players, level, pos, state);
                             } else {
                                 LevelChunkSection section = chunk.getSection(sectionIndex);
                                 ClientboundSectionBlocksUpdatePacket packet = new ClientboundSectionBlocksUpdatePacket(sectionPos, changedBlocks, section);
                                 this.broadcast(players, packet);
-                                packet.runUpdates((pos, state) -> this.broadcastBlockEntityIfNeeded(players, level, pos, state));
+                                // MultiPaper start - Sync chunk
+                                packet.runUpdates((pos, state) -> {
+                                    if (broadcastingToOtherServersOnly) this.broadcastBlockToPlayersLater(pos);
+                                    this.broadcastBlockEntityIfNeeded(players, level, pos, state);
+                                });
+                                // MultiPaper end - Sync chunk
                             }
                         }
                     }
@@ -327,14 +385,22 @@ public class ChunkHolder extends GenerationChunkHolder implements ca.spottedleaf
     private void broadcastBlockEntity(final List<ServerPlayer> players, final Level level, final BlockPos blockPos) {
         BlockEntity blockEntity = level.getBlockEntity(blockPos);
         if (blockEntity != null) {
-            Packet<?> packet = blockEntity.getUpdatePacket();
+            Packet<net.minecraft.network.protocol.game.ClientGamePacketListener> packet = blockEntity.getUpdatePacket(); // MultiPaper - Sync chunk
             if (packet != null) {
                 this.broadcast(players, packet);
+                if (!this.broadcastingToPlayersOnly) puregero.multipaper.MultiPaperChunkHandler.onBlockUpdate(newChunkHolder, net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket.create(blockEntity, BlockEntity::saveWithFullMetadata, false)); // MultiPaper - Sync chunk - sync the full block with external servers
             }
         }
     }
 
-    private void broadcast(final List<ServerPlayer> players, final Packet<?> packet) {
+    // MultiPaper start - Sync chunk
+    private void broadcast(final List<ServerPlayer> players, final Packet<? super net.minecraft.network.protocol.game.ClientGamePacketListener> packet) {
+        // we handle block entities separately
+        if (!this.broadcastingToPlayersOnly && !(packet instanceof net.minecraft.network.protocol.game.ClientboundBlockEntityDataPacket)) {
+            puregero.multipaper.MultiPaperChunkHandler.onBlockUpdate(this.newChunkHolder, packet);
+        }
+        if (this.broadcastingToOtherServersOnly) return;
+    // MultiPaper end - Sync chunk
         players.forEach(player -> player.connection.send(packet));
     }
 
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 075e294333115738b25b64a390df323276cf39f9..cbdb24dc0b62d05956a5666664f00fb44524f130 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -171,7 +171,23 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
         java.util.Objects.checkFromToIndex(0, size, raw.length);
         for (int i = 0; i < size; ++i) {
-            world.tickChunk(raw[i], randomTickSpeed);
+            // MultiPaper start - Sync chunk
+            LevelChunk chunk = raw[i];
+            // add temporary debug logs to check if a chunk can be not fully loaded
+            if (!chunk.loaded) {
+                LOGGER.warn("{} isn't loaded!", chunk);
+            }
+            if (this.getVisibleChunkIfPresent(chunk.getPos().longKey()) == null) {
+                LOGGER.warn("{} isn't in the visible chunks!", chunk);
+            }
+            // don't tick this chunk if we don't own it
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder holder = chunk.moonrise$getChunkHolder();
+            boolean isTicking = this.chunkMap.getDistanceManager().inEntityTickingRange(chunk.getPos().longKey());
+            if (isTicking && !holder.hasExternalLockRequest) puregero.multipaper.MultiPaper.lockChunk(holder);
+            if (!isTicking && holder.hasExternalLockRequest) puregero.multipaper.MultiPaper.unlockChunk(holder, chunk, holder.getEntityChunk());
+            if (!puregero.multipaper.MultiPaper.isChunkLocal(chunk)) continue;
+            world.tickChunk(chunk, randomTickSpeed);
+            // MultiPaper end - Sync chunk
 
             // call mid-tick tasks for chunk system
             if ((i & 7) == 0) {
@@ -658,6 +674,12 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         return this.chunkMap.size();
     }
 
+    // MultiPaper start - Sync chunk
+    public void addChunkHolderUnsafe(ChunkHolder holder) {
+        this.chunkHoldersToBroadcast.add(holder);
+    }
+    // MultiPaper end - Sync chunk
+
     public void blockChanged(final BlockPos pos) {
         int xc = SectionPos.blockToSectionCoord(pos.getX());
         int zc = SectionPos.blockToSectionCoord(pos.getZ());
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 19c3131f1b970aa0443800dcffd20b996ba8d7d5..dbc883141d7a6b653ef05746b5b458db2d3856dd 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -205,8 +205,10 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
     private final SleepStatus sleepStatus;
     private int emptyTime;
     private final PortalForcer portalForcer;
-    private final LevelTicks<Block> blockTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded);
-    private final LevelTicks<Fluid> fluidTicks = new LevelTicks<>(this::isPositionTickingWithEntitiesLoaded);
+    // MultiPaper start - Sync chunk
+    private final LevelTicks<Block> blockTicks = new LevelTicks<>(this, this::isPositionTickingWithEntitiesLoaded);
+    private final LevelTicks<Fluid> fluidTicks = new LevelTicks<>(this, this::isPositionTickingWithEntitiesLoaded);
+    // MultiPaper end - Sync chunk
     private final PathTypeCache pathTypesByPosCache = new PathTypeCache();
     private final Set<Mob> navigatingMobs = new ObjectOpenHashSet<>();
     private volatile boolean isUpdatingNavigations;
@@ -754,7 +756,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                 levelStorage.getDimensionPath(dimension).resolve("entities"),
                 server.forceSynchronousWrites()
             ),
-            this.chunkTaskScheduler
+            this.chunkTaskScheduler, this.getWorld().getName() // MultiPaper - Sync chunk
         );
         this.poiDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.PoiDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
         this.chunkDataController = new ca.spottedleaf.moonrise.patches.chunk_system.io.datacontroller.ChunkDataController((ServerLevel)(Object)this, this.chunkTaskScheduler);
@@ -1975,6 +1977,13 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             return;
         }
         // CraftBukkit end
+        // MultiPaper start - Sync chunk
+        // run neighbor updates later if we're handling block changes
+        if (puregero.multipaper.MultiPaperChunkHandler.blockUpdateChunk != null) {
+            io.papermc.paper.util.MCUtil.ensureMain(() -> this.updateNeighborsAt(pos, sourceBlock, ExperimentalRedstoneUtils.initialOrientation(this, null, null)));
+            return;
+        }
+        // MultiPaper end - Sync chunk
         if (captureBlockStates) { return; } // Paper - Cancel all physics during placement
         this.updateNeighborsAt(pos, sourceBlock, ExperimentalRedstoneUtils.initialOrientation(this, null, null));
     }
@@ -2432,6 +2441,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             this.server.updateEffectiveRespawnData();
         }
         // Paper end
+        if (!puregero.multipaper.externalserverprotocol.SpawnUpdatePacket.updatingSpawn) puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(this.uuid, new puregero.multipaper.externalserverprotocol.SpawnUpdatePacket(this)); // MultiPaper - Sync chunk
     }
 
     @Override
diff --git a/net/minecraft/world/entity/ai/village/poi/PoiManager.java b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
index 3b07e731e8d16c2b71cb55b85163029cf56fd02e..bb02f39ef1d5c1c1f2657afad6131d54e32f7ca1 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiManager.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiManager.java
@@ -103,7 +103,7 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
     }
 
     @Override
-    protected PoiSection getOrCreate(final long pos) {
+    public PoiSection getOrCreate(final long pos) { // MultiPaper - Sync chunk - make public
         final int chunkX = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionX(pos);
         final int chunkY = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionY(pos);
         final int chunkZ = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionZ(pos);
@@ -195,11 +195,16 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
     }
 
     public @Nullable PoiRecord add(final BlockPos pos, final Holder<PoiType> type) {
-        return this.getOrCreate(SectionPos.asLong(pos)).add(pos, type);
+        // MultiPaper start - Sync chunk
+        PoiRecord record = this.getOrCreate(SectionPos.asLong(pos)).add(pos, type);
+        puregero.multipaper.externalserverprotocol.SetPoiPacket.broadcastUpdate(this.world, pos);
+        return record;
+        // MultiPaper end - Sync chunk
     }
 
     public void remove(final BlockPos pos) {
         this.getOrLoad(SectionPos.asLong(pos)).ifPresent(poiSection -> poiSection.remove(pos));
+        puregero.multipaper.externalserverprotocol.SetPoiPacket.broadcastUpdate(this.world, pos); // MultiPaper - Sync chunk
     }
 
     public long getCountInRange(final Predicate<Holder<PoiType>> predicate, final BlockPos center, final int radius, final PoiManager.Occupancy occupancy) {
@@ -371,6 +376,7 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
         }
 
         record.acquireTicket();
+        puregero.multipaper.externalserverprotocol.SetPoiPacket.broadcastUpdate(world, record.getPos()); // MultiPaper - Sync chunk
         return Optional.of(record.getPos());
         // Paper end - optimise POI lookup
     }
@@ -401,9 +407,13 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
     }
 
     public boolean release(final BlockPos pos) {
-        return this.getOrLoad(SectionPos.asLong(pos))
+        boolean result = this.getOrLoad(SectionPos.asLong(pos)) // MultiPaper - Sync chunk
             .map(section -> section.release(pos))
             .orElseThrow(() -> Util.pauseInIde(new IllegalStateException("POI never registered at " + pos)));
+        // MultiPaper start - Sync chunk
+        puregero.multipaper.externalserverprotocol.SetPoiPacket.broadcastUpdate(this.world, pos);
+        return result;
+        // MultiPaper end - Sync chunk
     }
 
     public boolean exists(final BlockPos pos, final Predicate<Holder<PoiType>> predicate) {
@@ -439,6 +449,7 @@ public class PoiManager extends SectionStorage<PoiSection, PoiSection.Packed> im
 
     @Override
     public void setDirty(final long sectionPos) { // Paper - public
+        if (!puregero.multipaper.externalserverprotocol.SetPoiPacket.shouldSavePoi()) return; // MultiPaper - Sync chunk - don't save if modified by another server
         // Paper start - rewrite chunk system
         final int chunkX = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionX(sectionPos);
         final int chunkZ = ca.spottedleaf.moonrise.common.util.CoordinateUtils.getChunkSectionZ(sectionPos);
diff --git a/net/minecraft/world/entity/ai/village/poi/PoiRecord.java b/net/minecraft/world/entity/ai/village/poi/PoiRecord.java
index 5090221cdcd782582d56d2b5ef5a36640a610cd6..922b483279445c512077cec6060fdeabe870035a 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiRecord.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiRecord.java
@@ -36,6 +36,12 @@ public class PoiRecord {
         return this.freeTickets;
     }
 
+    // MultiPaper start - Sync chunk
+    public void setFreeTickets(int freeTickets) {
+        this.freeTickets = freeTickets;
+    }
+    // MultiPaper end - Sync chunk
+
     protected boolean acquireTicket() {
         if (this.freeTickets <= 0) {
             return false;
diff --git a/net/minecraft/world/entity/ai/village/poi/PoiSection.java b/net/minecraft/world/entity/ai/village/poi/PoiSection.java
index c708f4517ba1274db7046810044e69da4cc26b89..bffe0251ec5c37f3f8f67958da15d57f73d54644 100644
--- a/net/minecraft/world/entity/ai/village/poi/PoiSection.java
+++ b/net/minecraft/world/entity/ai/village/poi/PoiSection.java
@@ -129,7 +129,7 @@ public class PoiSection implements ca.spottedleaf.moonrise.patches.chunk_system.
         return this.getPoiRecord(pos).map(PoiRecord::getPoiType);
     }
 
-    private Optional<PoiRecord> getPoiRecord(final BlockPos pos) {
+    public Optional<PoiRecord> getPoiRecord(final BlockPos pos) { // MultiPaper - Sync chunk - make public
         return Optional.ofNullable(this.records.get(SectionPos.sectionRelativePos(pos)));
     }
 
diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index bad7d3df08b303aefc0d3de7ddd779c72e4bb136..606084c11b59ec51d2a5cfcf1a5e8a2a1744da04 100644
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -1205,7 +1205,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
                 this.setBlocksDirty(pos, oldState, newState);
             }
 
-            if ((updateFlags & Block.UPDATE_CLIENTS) != 0 && (!this.isClientSide() || (updateFlags & Block.UPDATE_INVISIBLE) == 0) && (this.isClientSide() || chunk == null || (chunk.getFullStatus() != null && chunk.getFullStatus().isOrAfter(FullChunkStatus.FULL)))) { // allow chunk to be null here as chunk.isReady() is false when we send our notification during block placement // Paper - rewrite chunk system - change from ticking to full
+            if ((true || (updateFlags & Block.UPDATE_CLIENTS) != 0) && (!this.isClientSide() || (updateFlags & Block.UPDATE_INVISIBLE) == 0) && (this.isClientSide() || chunk == null || (chunk.getFullStatus() != null && chunk.getFullStatus().isOrAfter(FullChunkStatus.FULL)))) { // allow chunk to be null here as chunk.isReady() is false when we send our notification during block placement // Paper - rewrite chunk system - change from ticking to full // MultiPaper - Sync chunk - send all block updates
                 this.sendBlockUpdated(pos, oldState, blockState, updateFlags);
             }
 
@@ -1550,6 +1550,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
             if (ticker.isRemoved()) {
                 toRemove.add(ticker); // Paper - Fix MC-117075 use removeAll
             } else if (tickBlockEntities && this.shouldTickBlocksAt(ticker.getPos())) {
+                if (!puregero.multipaper.MultiPaperChunkHandler.shouldTick(this, ticker.getPos())) continue; // MultiPaper - Sync chunk - should tick
                 ticker.tick();
                 // Paper start - rewrite chunk system
                 if ((++tickedEntities & 7) == 0) {
diff --git a/net/minecraft/world/level/LevelAccessor.java b/net/minecraft/world/level/LevelAccessor.java
index 700b9e7f141c4a2689690224c86376c9b957540e..49e4a8812974fdc663c73862c061b8142dfa6d4f 100644
--- a/net/minecraft/world/level/LevelAccessor.java
+++ b/net/minecraft/world/level/LevelAccessor.java
@@ -36,6 +36,32 @@ public interface LevelAccessor extends CommonLevelAccessor, ScheduledTickAccess
         return new ScheduledTick<>(type, pos, this.getGameTime() + tickDelay, this.nextSubTickCount());
     }
 
+    // MultiPaper start - Sync chunk
+    @Override
+    default void scheduleTick(BlockPos pos, Block block, int delay, TickPriority priority) {
+        puregero.multipaper.MultiPaperExternalBlocksHandler.onBlockScheduled(this.getMinecraftWorld(), pos);
+        ScheduledTickAccess.super.scheduleTick(pos, block, delay, priority);
+    }
+
+    @Override
+    default void scheduleTick(BlockPos pos, Block block, int delay) {
+        puregero.multipaper.MultiPaperExternalBlocksHandler.onBlockScheduled(this.getMinecraftWorld(), pos);
+        ScheduledTickAccess.super.scheduleTick(pos, block, delay);
+    }
+
+    @Override
+    default void scheduleTick(BlockPos pos, net.minecraft.world.level.material.Fluid fluid, int delay, TickPriority priority) {
+        puregero.multipaper.MultiPaperExternalBlocksHandler.onBlockScheduled(this.getMinecraftWorld(), pos);
+        ScheduledTickAccess.super.scheduleTick(pos, fluid, delay);
+    }
+
+    @Override
+    default void scheduleTick(BlockPos pos, net.minecraft.world.level.material.Fluid fluid, int delay) {
+        puregero.multipaper.MultiPaperExternalBlocksHandler.onBlockScheduled(this.getMinecraftWorld(), pos);
+        ScheduledTickAccess.super.scheduleTick(pos, fluid, delay);
+    }
+    // MultiPaper end - Sync chunk
+
     LevelData getLevelData();
 
     default long getGameTime() {
diff --git a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
index c7b3d34f4a44f0b7164ee35a6496d8357dae1c68..9aabb4f2c852e160e31153153cd69e907bf91307 100644
--- a/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BeaconBlockEntity.java
@@ -102,6 +102,7 @@ public class BeaconBlockEntity extends BlockEntity implements MenuProvider, Name
                 case 2:
                     BeaconBlockEntity.this.secondaryPower = BeaconBlockEntity.filterEffect(BeaconMenu.decodeEffect(value));
             }
+            BeaconBlockEntity.this.setChanged(); // MultiPaper - Sync chunk - send this block entity to other server
         }
 
         @Override
diff --git a/net/minecraft/world/level/block/entity/BlockEntity.java b/net/minecraft/world/level/block/entity/BlockEntity.java
index d4453e8236daf2c540478c051f3a4ec2f7e90f7e..d35b3865b9350c48e183b49aa8c03c3fb06392d4 100644
--- a/net/minecraft/world/level/block/entity/BlockEntity.java
+++ b/net/minecraft/world/level/block/entity/BlockEntity.java
@@ -225,6 +225,7 @@ public abstract class BlockEntity implements DebugValueSource, TypedInstance<Blo
     }
 
     protected static void setChanged(final Level level, final BlockPos worldPosition, final BlockState blockState) {
+        ((ServerLevel) level).getChunkSource().blockChanged(worldPosition); // MultiPaper - Sync chunk
         level.blockEntityChanged(worldPosition);
         if (!blockState.isAir()) {
             level.updateNeighbourForOutputSignal(worldPosition, blockState.getBlock());
diff --git a/net/minecraft/world/level/block/entity/ComparatorBlockEntity.java b/net/minecraft/world/level/block/entity/ComparatorBlockEntity.java
index fea412b0c2afc63da5037b49ee4e245b829943fd..690a2600160e924def584e2a82e7b913673f5817 100644
--- a/net/minecraft/world/level/block/entity/ComparatorBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/ComparatorBlockEntity.java
@@ -31,5 +31,6 @@ public class ComparatorBlockEntity extends BlockEntity {
 
     public void setOutputSignal(final int value) {
         this.output = value;
+        this.setChanged(); // MultiPaper - Sync chunk
     }
 }
diff --git a/net/minecraft/world/level/block/piston/PistonBaseBlock.java b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
index c02394b5ea59c9f3b5d2b4f40b4f69d701698aed..d4ba4093a0fd739c55dacb6b0eb4f0516444e31c 100644
--- a/net/minecraft/world/level/block/piston/PistonBaseBlock.java
+++ b/net/minecraft/world/level/block/piston/PistonBaseBlock.java
@@ -385,6 +385,11 @@ public class PistonBaseBlock extends DirectionalBlock {
             toUpdate[updateIndex++] = state;
         }
 
+        // MultiPaper start - Sync chunk
+        java.util.HashSet<net.minecraft.world.level.ChunkPos> takeControlOfChunks = new java.util.HashSet<>();
+        takeControlOfChunks.add(net.minecraft.world.level.ChunkPos.containing(pistonPos));
+        // MultiPaper end - Sync chunk
+
         for (int i = toPush.size() - 1; i >= 0; i--) {
             // Paper start - fix a variety of piston desync dupes
             boolean allowDesync = io.papermc.paper.configuration.GlobalConfiguration.get().unsupportedSettings.allowPistonDuplication;
@@ -395,23 +400,22 @@ public class PistonBaseBlock extends DirectionalBlock {
             pos = pos.relative(pushDirection);
             deleteAfterMove.remove(pos);
             BlockState actualState = Blocks.MOVING_PISTON.defaultBlockState().setValue(FACING, direction);
-            level.setBlock(pos, actualState, Block.UPDATE_NONE | Block.UPDATE_MOVE_BY_PISTON);
+            takeControlOfChunks.add(net.minecraft.world.level.ChunkPos.containing(pos)); // MultiPaper - Sync chunk
             // Paper start - fix a variety of piston desync dupes
             if (!allowDesync) {
                 blockState = level.getBlockState(oldPos);
                 deleteAfterMove.replace(oldPos, blockState);
             }
-            level.setBlockEntity(
-                MovingPistonBlock.newMovingBlockEntity(pos, actualState, allowDesync ? toPushShapes.get(i) : blockState, direction, extending, false)
-            );
+            puregero.multipaper.externalserverprotocol.PistonMoveBlockStartPacket.startBlockMove((ServerLevel) level, oldPos, pos, direction, extending, false); // MultiPaper - Sync chunk
             if (!allowDesync) {
-                level.setBlock(oldPos, Blocks.AIR.defaultBlockState(), Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE | Block.UPDATE_MOVE_BY_PISTON | Block.UPDATE_SKIP_ON_PLACE); // set air to prevent later physics updates from seeing this block
+                // level.setBlock(oldPos, Blocks.AIR.defaultBlockState(), Block.UPDATE_CLIENTS | Block.UPDATE_KNOWN_SHAPE | Block.UPDATE_MOVE_BY_PISTON | Block.UPDATE_SKIP_ON_PLACE); // set air to prevent later physics updates from seeing this block // MultiPaper - Sync chunk
             }
             // Paper end - fix a variety of piston desync dupes
             toUpdate[updateIndex++] = blockState;
         }
 
         if (extending) {
+            takeControlOfChunks.add(net.minecraft.world.level.ChunkPos.containing(armPos)); // MultiPaper - Sync chunk
             PistonType type = this.isSticky ? PistonType.STICKY : PistonType.DEFAULT;
             BlockState state = Blocks.PISTON_HEAD.defaultBlockState().setValue(PistonHeadBlock.FACING, direction).setValue(PistonHeadBlock.TYPE, type);
             BlockState blockState = Blocks.MOVING_PISTON
@@ -421,8 +425,10 @@ public class PistonBaseBlock extends DirectionalBlock {
             deleteAfterMove.remove(armPos);
             level.setBlock(armPos, blockState, Block.UPDATE_NONE | Block.UPDATE_MOVE_BY_PISTON);
             level.setBlockEntity(MovingPistonBlock.newMovingBlockEntity(armPos, blockState, state, direction, true, true));
+            ((ServerLevel) level).getChunkSource().blockChanged(armPos); // MultiPaper - Sync chunk
         }
 
+        takeControlOfChunks((ServerLevel) level, takeControlOfChunks); // MultiPaper - Sync chunk
         BlockState air = Blocks.AIR.defaultBlockState();
 
         for (BlockPos pos : deleteAfterMove.keySet()) {
@@ -459,9 +465,41 @@ public class PistonBaseBlock extends DirectionalBlock {
             level.updateNeighborsAt(armPos, Blocks.PISTON_HEAD, orientation);
         }
 
+        // MultiPaper start - Sync chunk
+        // broadcast changes to sync other servers
+        net.minecraft.world.level.chunk.LevelChunk chunk = level.getChunkIfLoaded(pistonPos);
+        if (chunk != null) {
+            chunk.moonrise$getChunkHolder().vanillaChunkHolder.broadcastChangesToOtherServers(chunk);
+        }
+
+        if (extending) {
+            net.minecraft.world.level.chunk.LevelChunk facingChunk = level.getChunkIfLoaded(armPos);
+            if (facingChunk != null) {
+                facingChunk.moonrise$getChunkHolder().vanillaChunkHolder.broadcastChangesToOtherServers(facingChunk);
+            }
+        }
+        // MultiPaper end - Sync chunk
+
         return true;
     }
 
+    // MultiPaper start - Sync chunk
+    private static void takeControlOfChunks(ServerLevel level, java.util.HashSet<net.minecraft.world.level.ChunkPos> takeControlOfChunks) {
+        if (takeControlOfChunks.stream().anyMatch(pos -> puregero.multipaper.MultiPaper.isChunkExternal(puregero.multipaper.MultiPaper.getChunkHolder(level, pos.x(), pos.z())))
+                && takeControlOfChunks.stream().allMatch(pos -> puregero.multipaper.MultiPaper.getChunkHolder(level, pos.x(), pos.z()).hasExternalLockRequest)) {
+            puregero.multipaper.MultiPaper.getConnection().sendAndAwaitReply(
+                    new puregero.multipaper.mastermessagingprotocol.messages.masterbound.RequestChunkOwnershipMessage(
+                            level.getWorld().getName(),
+                            takeControlOfChunks.stream()
+                                    .map(pos -> new puregero.multipaper.mastermessagingprotocol.ChunkKey(level.getWorld().getName(), pos.x(), pos.z()))
+                                    .toArray(puregero.multipaper.mastermessagingprotocol.ChunkKey[]::new)
+                    ),
+                    puregero.multipaper.mastermessagingprotocol.messages.serverbound.BooleanMessageReply.class
+            );
+        }
+    }
+    // MultiPaper end - Sync chunk
+
     @Override
     protected BlockState rotate(final BlockState state, final Rotation rotation) {
         return state.setValue(FACING, rotation.rotate(state.getValue(FACING)));
diff --git a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
index 3fcb1a45a547681740d6ec8946b274539448cf95..41661c9dc908eb9805c40479b96596a78dafd7f9 100644
--- a/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
+++ b/net/minecraft/world/level/block/piston/PistonMovingBlockEntity.java
@@ -37,7 +37,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
     private static final float DEFAULT_PROGRESS = 0.0F;
     private static final boolean DEFAULT_EXTENDING = false;
     private static final boolean DEFAULT_SOURCE = false;
-    private BlockState movedState = DEFAULT_BLOCK_STATE;
+    public BlockState movedState = DEFAULT_BLOCK_STATE; // MultiPaper - Sync chunk - make public
     private Direction direction = Direction.DOWN; // Paper - default to first value to avoid NPE
     private boolean extending = false;
     private boolean isSourcePiston = false;
@@ -280,7 +280,7 @@ public class PistonMovingBlockEntity extends BlockEntity {
     }
 
     public void finalTick() {
-        if (this.level != null && (this.progressO < 1.0F || this.level.isClientSide())) {
+        if (this.level != null && (this.progressO < 1.0F || this.level.isClientSide()) && !puregero.multipaper.MultiPaper.isChunkExternal((net.minecraft.server.level.ServerLevel) this.level, getBlockPos())) { // MultiPaper - Sync chunk
             this.progress = 1.0F;
             this.progressO = this.progress;
             this.level.removeBlockEntity(this.worldPosition);
@@ -343,6 +343,8 @@ public class PistonMovingBlockEntity extends BlockEntity {
             if (entity.progress >= 1.0F) {
                 entity.progress = 1.0F;
             }
+
+            entity.setChanged(); // MultiPaper - Sync chunk
         }
     }
 
diff --git a/net/minecraft/world/level/chunk/LevelChunk.java b/net/minecraft/world/level/chunk/LevelChunk.java
index 681563230061212afa7994c8eb16c068754db5dd..d920e5dacbb5c078a76cf80c6f3a7a258584ffbf 100644
--- a/net/minecraft/world/level/chunk/LevelChunk.java
+++ b/net/minecraft/world/level/chunk/LevelChunk.java
@@ -86,13 +86,17 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         }
     };
     private final Map<BlockPos, LevelChunk.RebindableTickingBlockEntityWrapper> tickersInLevel = Maps.newHashMap();
-    private boolean loaded;
+    public boolean loaded; // MultiPaper - Sync chunk - make public
     private final ServerLevel level; // CraftBukkit - type
     private @Nullable Supplier<FullChunkStatus> fullStatus;
     private LevelChunk.@Nullable PostLoadProcessor postLoad;
     private final Int2ObjectMap<GameEventListenerRegistry> gameEventListenerRegistrySections;
-    private final LevelChunkTicks<Block> blockTicks;
-    private final LevelChunkTicks<Fluid> fluidTicks;
+    // MultiPaper start - Sync chunk
+    public LevelChunkTicks<Block> blockTicks;
+    public LevelChunkTicks<Fluid> fluidTicks;
+    public net.minecraft.nbt.@Nullable ListTag entitiesToLoad;
+    public net.minecraft.nbt.@Nullable ListTag blockEntitiesToLoad;
+    // MultiPaper end - Sync chunk
     private LevelChunk.UnsavedListener unsavedListener = chunkPos -> {};
     // CraftBukkit start
     public boolean mustNotSave;
@@ -403,8 +407,9 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         boolean blockChanged = !oldState.is(newBlock);
         boolean movedByPiston = (flags & Block.UPDATE_MOVE_BY_PISTON) != 0;
         boolean sideEffects = (flags & Block.UPDATE_SKIP_BLOCK_ENTITY_SIDEEFFECTS) == 0;
+        boolean canOnRemove = puregero.multipaper.MultiPaperChunkHandler.blockUpdateChunk == null; // MultiPaper - Sync chunk
         if (blockChanged && oldState.hasBlockEntity() && !state.shouldChangedStateKeepBlockEntity(oldState)) {
-            if (!this.level.isClientSide() && sideEffects) {
+            if (canOnRemove && !this.level.isClientSide() && sideEffects) { // MultiPaper - Sync chunk - only call preRemoveSideEffects when we're not updating a chunk
                 BlockEntity blockEntity = this.level.getBlockEntity(pos);
                 if (blockEntity != null) {
                     blockEntity.preRemoveSideEffects(pos, oldState);
@@ -414,7 +419,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
             this.removeBlockEntity(pos);
         }
 
-        if ((blockChanged || newBlock instanceof BaseRailBlock)
+        if (canOnRemove && (blockChanged || newBlock instanceof BaseRailBlock) // MultiPaper - Sync chunk - only call affectNeighborsAfterRemoval when we're not updating a chunk
             && this.level instanceof ServerLevel serverLevel
             && ((flags & Block.UPDATE_NEIGHBORS) != 0 || movedByPiston)) {
             oldState.affectNeighborsAfterRemoval(serverLevel, pos, movedByPiston);
@@ -578,6 +583,12 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     @Override
     public void removeBlockEntity(final BlockPos pos) {
+        // MultiPaper start - Sync chunk
+        // allow removing a block entity on unloaded block entities
+        if (this.blockEntitiesToLoad != null) {
+            this.blockEntitiesToLoad.removeIf(tag -> pos.equals(BlockEntity.getPosFromTag(ChunkPos.containing(pos), (CompoundTag) tag)));
+        }
+        // MultiPaper end - Sync chunk
         if (this.isInLevel()) {
             BlockEntity removeThis = this.blockEntities.remove(pos);
             // CraftBukkit start - SPIGOT-5561: Also remove from pending map
@@ -630,11 +641,13 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
     // CraftBukkit start
     public void loadCallback() {
         if (this.loadedTicketLevel) { LOGGER.error("Double calling chunk load!", new Throwable()); } // Paper
+        this.mustNotSave = false; // MultiPaper - Sync chunk - reset mustNotSave on chunk load
         // Paper start
         this.loadedTicketLevel = true;
         // Paper end
         org.bukkit.Server server = this.level.getCraftServer();
         // Paper - rewrite chunk system
+        puregero.multipaper.MultiPaperChunkHandler.onChunkLoad(this); // MultiPaper - Sync chunk
         if (server != null) {
             /*
              * If it's a new world, the first few chunks are generated inside
@@ -686,6 +699,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
 
     @Override
     public boolean isUnsaved() {
+        if (puregero.multipaper.MultiPaper.isChunkExternal(this)) return false; // MultiPaper - Sync chunk - only save if no one else owns it
         // Paper start - rewrite chunk system
         final long gameTime = this.level.getGameTime();
         if (((ca.spottedleaf.moonrise.patches.chunk_system.ticks.ChunkSystemLevelChunkTicks)this.blockTicks).moonrise$isDirty(gameTime)
@@ -751,7 +765,7 @@ public class LevelChunk extends ChunkAccess implements DebugValueSource, ca.spot
         this.loaded = loaded;
     }
 
-    public Level getLevel() {
+    public ServerLevel getLevel() { // MultiPaper - Sync chunk
         return this.level;
     }
 
diff --git a/net/minecraft/world/level/chunk/ProtoChunk.java b/net/minecraft/world/level/chunk/ProtoChunk.java
index 93a14a457312ee0b4ddf6e80c425f1a50cc9edf1..daf81ad763d0573384dff64469b5aa2ad1e3516d 100644
--- a/net/minecraft/world/level/chunk/ProtoChunk.java
+++ b/net/minecraft/world/level/chunk/ProtoChunk.java
@@ -47,8 +47,10 @@ public class ProtoChunk extends ChunkAccess {
     private final List<CompoundTag> entities = Lists.newArrayList();
     private @Nullable CarvingMask carvingMask;
     private @Nullable BelowZeroRetrogen belowZeroRetrogen;
-    private final ProtoChunkTicks<Block> blockTicks;
-    private final ProtoChunkTicks<Fluid> fluidTicks;
+    // MultiPaper start - Sync chunk - make public and non-final
+    public ProtoChunkTicks<Block> blockTicks;
+    public ProtoChunkTicks<Fluid> fluidTicks;
+    // MultiPaper end - Sync chunk - make public and non-final
 
     public ProtoChunk(
         final ChunkPos chunkPos,
diff --git a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
index d00c0cd57fec367acc17c15f247e22990c5bdf7c..350648b5f44bd7200c00ad98005b081695b43630 100644
--- a/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
+++ b/net/minecraft/world/level/chunk/storage/RegionFileStorage.java
@@ -22,6 +22,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     private final Long2ObjectLinkedOpenHashMap<RegionFile> regionCache = new Long2ObjectLinkedOpenHashMap<>();
     private final RegionStorageInfo info;
     private final Path folder;
+    public final String folderStr; // MultiPaper - Sync chunk
     private final boolean sync;
 
     // Paper start - recalculate region file headers
@@ -268,6 +269,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
 
     public RegionFileStorage(final RegionStorageInfo info, final Path folder, final boolean sync) {
         this.folder = folder;
+        this.folderStr = folder.toString(); // MultiPaper - Sync chunk
         this.sync = sync;
         this.info = info;
         this.isChunkData = info.dfuType()[0] == net.minecraft.util.datafix.DataFixTypes.CHUNK; // Paper - recalculate region file headers
@@ -345,6 +347,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     // Paper end
 
     public @Nullable CompoundTag read(final ChunkPos pos) throws IOException {
+        if (true) throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
         // CraftBukkit start - SPIGOT-5680: There's no good reason to preemptively create files on read, save that for writing
         RegionFile region = this.getRegionFile(pos, true);
         if (region == null) {
@@ -397,6 +400,7 @@ public class RegionFileStorage implements AutoCloseable, ca.spottedleaf.moonrise
     }
 
     public void write(final ChunkPos pos, final @Nullable CompoundTag value) throws IOException {
+        if (true) throw new UnsupportedOperationException(); // MultiPaper - Sync chunk
         if (!SharedConstants.DEBUG_DONT_SAVE_WORLD) {
             RegionFile region = this.getRegionFile(pos, value == null); // CraftBukkit // Paper - rewrite chunk system
             // Paper start - rewrite chunk system
diff --git a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
index 7d29fdc6548324d97ec34d476b5b48ec0bdba0d2..e6ab23855b7b4a8ddf4ca74a62f6c26a095e4e40 100644
--- a/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
+++ b/net/minecraft/world/level/chunk/storage/SerializableChunkData.java
@@ -371,6 +371,11 @@ public record SerializableChunkData(
                 postLoadChunk(level, this.entities, this.blockEntities),
                 BlendingData.unpack(this.blendingData)
             );
+            // MultiPaper start - Sync chunk
+            // keep copy of entities and block entities until they're loaded
+            ((LevelChunk) chunk).entitiesToLoad = new ListTag(new ArrayList<>(this.entities));
+            ((LevelChunk) chunk).blockEntitiesToLoad = new ListTag(new ArrayList<>(this.blockEntities));
+            // MultiPaper end - Sync chunk
         } else {
             ProtoChunkTicks<Block> blockTicks = ProtoChunkTicks.load(this.packedTicks.blocks());
             ProtoChunkTicks<Fluid> fluidTicks = ProtoChunkTicks.load(this.packedTicks.fluids());
@@ -679,6 +684,12 @@ public record SerializableChunkData(
                         }
                     }
                 }
+
+                // MultiPaper start - Sync chunk
+                // keep copy of entities and block entities until they're loaded
+                levelChunk.entitiesToLoad = null;
+                levelChunk.blockEntitiesToLoad = null;
+                // MultiPaper end - Sync chunk
             };
     }
 
diff --git a/net/minecraft/world/level/levelgen/PhantomSpawner.java b/net/minecraft/world/level/levelgen/PhantomSpawner.java
index ed227f3bcc0a9efe34d8fc32d10d7c7029c3dd5b..b96b5f550313e2765ffbd79a912fbffc3ad64369 100644
--- a/net/minecraft/world/level/levelgen/PhantomSpawner.java
+++ b/net/minecraft/world/level/levelgen/PhantomSpawner.java
@@ -52,6 +52,7 @@ public class PhantomSpawner implements CustomSpawner {
                                             BlockPos spawnPos = playerPos.above(20 + random.nextInt(15))
                                                 .east(-10 + random.nextInt(21))
                                                 .south(-10 + random.nextInt(21));
+                                            if (!level.hasChunkAt(spawnPos)) continue; // MultiPaper - Sync chunk - don't load chunks when spawning phantoms
                                             BlockState blockState = level.getBlockState(spawnPos);
                                             FluidState fluidState = level.getFluidState(spawnPos);
                                             if (NaturalSpawner.isValidEmptySpawnBlock(level, spawnPos, blockState, fluidState, EntityTypes.PHANTOM)) {
diff --git a/net/minecraft/world/level/portal/PortalForcer.java b/net/minecraft/world/level/portal/PortalForcer.java
index a468a1013058089115d023e8957f976ae34935ad..8eebc6c8c85f56d62b870072ebf3e40589ebb4af 100644
--- a/net/minecraft/world/level/portal/PortalForcer.java
+++ b/net/minecraft/world/level/portal/PortalForcer.java
@@ -64,7 +64,7 @@ public class PortalForcer {
                     }
                     // Paper end - Configurable nether ceiling damage
 
-                    net.minecraft.world.level.chunk.ChunkAccess lowest = this.level.getChunk(pos.getX() >> 4, pos.getZ() >> 4, net.minecraft.world.level.chunk.status.ChunkStatus.EMPTY);
+                    net.minecraft.world.level.chunk.ChunkAccess lowest = this.level.getChunk(pos.getX() >> 4, pos.getZ() >> 4, net.minecraft.world.level.chunk.status.ChunkStatus.FULL); // MultiPaper - Sync chunk - for some reason empty was resetting the chunks nearby the portal
                     net.minecraft.world.level.levelgen.BelowZeroRetrogen belowZeroRetrogen;
                     if (!lowest.getPersistedStatus().isOrAfter(net.minecraft.world.level.chunk.status.ChunkStatus.FULL)
                             // check below zero retrogen so that pre 1.17 worlds still load portals (JMP)
diff --git a/net/minecraft/world/ticks/LevelTicks.java b/net/minecraft/world/ticks/LevelTicks.java
index 7bf9e93692debcef511bfa316aeef32ae8a7d3b7..7754a81282137206600ff3b73eb392d5288c9700 100644
--- a/net/minecraft/world/ticks/LevelTicks.java
+++ b/net/minecraft/world/ticks/LevelTicks.java
@@ -43,7 +43,11 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
         }
     };
 
-    public LevelTicks(final LongPredicate tickCheck) {
+    // MultiPaper start - Sync chunk
+    public net.minecraft.server.level.ServerLevel level;
+    public LevelTicks(final net.minecraft.server.level.ServerLevel level, final LongPredicate tickCheck) {
+        this.level = level;
+    // MultiPaper end - Sync chunk
         this.tickCheck = tickCheck;
     }
 
@@ -108,7 +112,7 @@ public class LevelTicks<T> implements LevelTickAccess<T> {
                 LevelChunkTicks<T> candidateContainer = this.allContainers.get(chunkPos);
                 if (candidateContainer == null) {
                     it.remove();
-                } else {
+                } else if (candidateContainer.count() <= 0 || puregero.multipaper.MultiPaper.isChunkLocal(level.moonrise$getChunkTaskScheduler().chunkHolderManager.getChunkHolder(chunkPos))) { // MultiPaper - Sync chunk - don't tick chunks that we don't own
                     ScheduledTick<T> scheduledTick = candidateContainer.peek();
                     if (scheduledTick == null) {
                         it.remove();
