From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koyorice <87787929+Koyorice@users.noreply.github.com>
Date: Sat, 4 Oct 2025 23:35:33 +0800
Subject: [PATCH] Sync entity


diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
index 5b7ed8e5067683143c99473a126e17baa531903c..2fda07f8e4f2b84808574b779f245565794477ad 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/ChunkEntitySlices.java
@@ -51,7 +51,7 @@ public final class ChunkEntitySlices {
     private final EntityCollectionBySection hardCollidingEntities;
     private final Reference2ObjectOpenHashMap<Class<? extends Entity>, EntityCollectionBySection> entitiesByClass;
     private final Reference2ObjectOpenHashMap<EntityType<?>, EntityCollectionBySection> entitiesByType;
-    private final EntityList entities = new EntityList();
+    public final EntityList entities = new EntityList(); // MultiPaper - Sync entity - make public
 
     public FullChunkStatus status;
     public final ChunkData chunkData;
@@ -193,6 +193,23 @@ public final class ChunkEntitySlices {
                     }
                 }
             }
+            // MultiPaper start - Sync entity
+            // clear external player vehicles
+            if (puregero.multipaper.MultiPaperEntitiesHandler.getControllingPassenger(entity) != null) {
+                java.util.Queue<Entity> toRemove = new java.util.LinkedList<>();
+                Entity toRemoveEntity = entity.getRootVehicle();
+                do {
+                    toRemove.addAll(toRemoveEntity.getPassengers());
+                    toRemoveEntity.forceEjectPassengers();
+                    if (!(toRemoveEntity instanceof net.minecraft.server.level.ServerPlayer)) {
+                        toRemoveEntity.setRemoved(Entity.RemovalReason.UNLOADED_TO_CHUNK);
+                    }
+                } while ((toRemoveEntity = toRemove.poll()) != null);
+            }
+            if (entity instanceof puregero.multipaper.ExternalPlayer) {
+                this.removeEntity(entity, entity.moonrise$getSectionY());
+            }
+            // MultiPaper end - Sync entity
         }
 
         return this.entities.size() != 0;
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
index 328784ba08236553e49d62320b79f7222292412f..c024c61a34ee66133a73019023ad9b4052f7a9e0 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -97,6 +97,17 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         return visibility.isAccessible() ? entity : null;
     }
 
+    // MultiPaper start - Sync entity
+    @javax.annotation.Nullable
+    public Entity getEntityIgnoringAccessible(UUID uuid) {
+        return this.entityByUUID.get(uuid);
+    }
+
+    public boolean isChunkLoaded(ChunkPos chunkPos) {
+        return this.getChunk(chunkPos.x(), chunkPos.z()) != null;
+    }
+    // MultiPaper end - Sync entity
+
     @Override
     public Entity get(final int id) {
         return maskNonAccessible(this.entityById.get((long)id));
@@ -443,8 +454,8 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         ((ChunkSystemEntity)entity).moonrise$setSectionX(sectionX);
         ((ChunkSystemEntity)entity).moonrise$setSectionY(sectionY);
         ((ChunkSystemEntity)entity).moonrise$setSectionZ(sectionZ);
-        final ChunkEntitySlices slices = this.getOrCreateChunk(sectionX, sectionZ);
-        if (!slices.addEntity(entity, sectionY)) {
+        final ChunkEntitySlices slices = entity instanceof puregero.multipaper.ExternalPlayer ? this.getChunk(sectionX, sectionZ) : this.getOrCreateChunk(sectionX, sectionZ); // MultiPaper - Sync entity - don't create the chunk for external players
+        if (slices != null && !slices.addEntity(entity, sectionY)) { // MultiPaper - Sync entity - don't create the chunk for external players
             LOGGER.warn("Entity " + entity + " added to world '" + WorldUtil.getWorldName(this.world) + "', but was already contained in entity chunk (" + sectionX + "," + sectionZ + ")");
         }
 
@@ -479,13 +490,13 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         final ChunkEntitySlices slices = this.getChunk(sectionX, sectionZ);
         // all entities should be in a chunk
         if (slices == null) {
-            LOGGER.warn("Cannot remove entity " + entity + " from null entity slices (" + sectionX + "," + sectionZ + ")");
+            if (!(entity instanceof puregero.multipaper.ExternalPlayer)) LOGGER.warn("Cannot remove entity " + entity + " from null entity slices (" + sectionX + "," + sectionZ + ")"); // MultiPaper - Sync entity - don't create the chunk for external players
         } else {
             if (slices.isPreventingStatusUpdates()) {
                 throw new IllegalStateException("Attempting to remove entity " + entity + " from entity slices (" + sectionX + "," + sectionZ + ") that is receiving status updates");
             }
             if (!slices.removeEntity(entity, sectionY)) {
-                LOGGER.warn("Failed to remove entity " + entity + " from entity slices (" + sectionX + "," + sectionZ + ")");
+                if (!(entity instanceof puregero.multipaper.ExternalPlayer)) LOGGER.warn("Failed to remove entity " + entity + " from entity slices (" + sectionX + "," + sectionZ + ")"); // MultiPaper - Sync entity - don't create the chunk for external players
             }
         }
         ((ChunkSystemEntity)entity).moonrise$setSectionX(Integer.MIN_VALUE);
@@ -541,13 +552,13 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         this.checkThread(sectionX, sectionZ, "Cannot move entity off-main");
 
         final ChunkEntitySlices old = this.getChunk(sectionX, sectionZ);
-        final ChunkEntitySlices slices = this.getOrCreateChunk(newSectionX, newSectionZ);
+        final ChunkEntitySlices slices = entity instanceof puregero.multipaper.ExternalPlayer ? this.getChunk(newSectionX, newSectionZ) : this.getOrCreateChunk(newSectionX, newSectionZ); // MultiPaper - Sync entity - don't create the chunk for external players
 
-        if (!old.removeEntity(entity, sectionY)) {
-            LOGGER.warn("Could not remove entity " + entity + " from its old chunk section (" + sectionX + "," + sectionY + "," + sectionZ + ") since it was not contained in the section");
+        if (old != null && !old.removeEntity(entity, sectionY)) { // MultiPaper - Sync entity - don't create the chunk for external players
+            if (!(entity instanceof puregero.multipaper.ExternalPlayer)) LOGGER.warn("Could not remove entity " + entity + " from its old chunk section (" + sectionX + "," + sectionY + "," + sectionZ + ") since it was not contained in the section"); // MultiPaper - Sync entity - don't create the chunk for external players
         }
 
-        if (!slices.addEntity(entity, newSectionY)) {
+        if (slices != null && !slices.addEntity(entity, newSectionY)) { // MultiPaper - Sync entity - don't create the chunk for external players
             LOGGER.warn("Could not add entity " + entity + " to its new chunk section (" + newSectionX + "," + newSectionY + "," + newSectionZ + ") as it is already contained in the section");
         }
 
@@ -555,7 +566,7 @@ public abstract class EntityLookup implements LevelEntityGetter<Entity> {
         ((ChunkSystemEntity)entity).moonrise$setSectionY(newSectionY);
         ((ChunkSystemEntity)entity).moonrise$setSectionZ(newSectionZ);
 
-        if (old.isEmpty()) {
+        if (old != null && old.isEmpty()) { // MultiPaper - Sync entity - don't create the chunk for external players
             this.onEmptySlices(sectionX, sectionZ);
         }
 
diff --git a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
index 203447556dda19df6a6ea812675253362cf12d45..bf65eb283c031747329ce88a6e5f3ff38357847f 100644
--- a/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/scheduling/NewChunkHolder.java
@@ -74,6 +74,7 @@ public final class NewChunkHolder {
     public boolean hasExternalLockRequest = false; // Have we sent a request to lock this chunk?
     public java.util.Set<puregero.multipaper.ExternalServer> externalSubscribers = java.util.concurrent.ConcurrentHashMap.newKeySet(); // Chunk subscribers
     // MultiPaper end - Sync chunk
+    public java.util.Set<puregero.multipaper.ExternalServer> externalEntitiesSubscribers = java.util.concurrent.ConcurrentHashMap.newKeySet(); // MultiPaper - Sync entity - Entities subscribers
 
     // load/unload state
 
@@ -167,6 +168,11 @@ public final class NewChunkHolder {
             if (result != null) {
                 this.entityDataLoadTaskWaiters = null;
                 this.pendingEntityChunk = result.left() == null ? EMPTY_ENTITY_CHUNK : result.left();
+                // MultiPaper start - Sync entity
+                // We can't leave entities cached in nbt
+                // We need all entities accessible in case another server updates the entities in the chunk
+                puregero.multipaper.MultiPaper.runSync(() -> world.moonrise$getChunkTaskScheduler().chunkHolderManager.getOrCreateEntityChunk(chunkX, chunkZ, false));
+                // MultiPaper end - Sync entity
                 if (result.right() != null) {
                     LOGGER.error("Unhandled entity data load exception, data data will be lost: ", result.right());
                 }
@@ -945,6 +951,7 @@ public final class NewChunkHolder {
             // we need to delay the callback until after determining transience, otherwise a potential loader could
             // set entityChunk before we do
             this.asyncCompleteUnloadTask(MoonriseRegionFileIO.RegionFileType.ENTITY_DATA, entityWrite[0]);
+            puregero.multipaper.MultiPaperEntitiesHandler.onEntitiesUnload(this); // MultiPaper - Sync entity
         }
 
         // unload poi data
@@ -1818,6 +1825,7 @@ public final class NewChunkHolder {
                     }
                 });
 
+                puregero.multipaper.MultiPaperEntitiesHandler.willWriteEntities(this); // MultiPaper - Sync entity
                 MoonriseRegionFileIO.scheduleSave(
                     this.world, this.chunkX, this.chunkZ, toWrite, null, MoonriseRegionFileIO.RegionFileType.ENTITY_DATA, Priority.NORMAL
                 );
@@ -1834,6 +1842,7 @@ public final class NewChunkHolder {
                 return false;
             }
 
+            puregero.multipaper.MultiPaperEntitiesHandler.willWriteEntities(this); // MultiPaper - Sync entity
             MoonriseRegionFileIO.scheduleSave(this.world, this.chunkX, this.chunkZ, save, MoonriseRegionFileIO.RegionFileType.ENTITY_DATA);
             this.lastEntitySaveNull = save == null;
             if (entitySave != null) {
diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 9a26b6ecf21f53c7e00fc8b1b1d907efd51cd6c2..29e1e2fbd7c7cc3b91e97305c0d51dc0c4de48f8 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -1345,7 +1345,17 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayers(final Packet<? super ClientGamePacketListener> packet) {
+            // MultiPaper start - Sync entity
+            // are we ticking this entity?
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder newChunkHolder = puregero.multipaper.MultiPaper.getChunkHolder(entity);
+            if (!puregero.multipaper.MultiPaper.isExternalPlayer(puregero.multipaper.MultiPaperEntitiesHandler.getControllingPassenger(entity)) &&
+                    (puregero.multipaper.MultiPaper.isRealPlayer(puregero.multipaper.MultiPaperEntitiesHandler.getControllingPassenger(entity)) ||
+                            puregero.multipaper.MultiPaper.isChunkLocal(newChunkHolder))) {
+                puregero.multipaper.MultiPaperEntitiesHandler.onEntityUpdate(entity, packet, newChunkHolder);
+            }
+            // MultiPaper end - Sync entity
             for (ServerPlayerConnection connection : this.seenBy) {
+                if (puregero.multipaper.MultiPaper.isExternalPlayer(connection.getPlayer())) continue; // MultiPaper - Sync entity - don't send to external players
                 connection.send(packet);
             }
         }
@@ -1360,8 +1370,18 @@ public class ChunkMap extends SimpleRegionStorage implements ChunkHolder.PlayerP
 
         @Override
         public void sendToTrackingPlayersFiltered(final Packet<? super ClientGamePacketListener> packet, final Predicate<ServerPlayer> targetPredicate) {
+            // MultiPaper start - Sync entity
+            // are we ticking this entity?
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder newChunkHolder = puregero.multipaper.MultiPaper.getChunkHolder(entity);
+            if (!puregero.multipaper.MultiPaper.isExternalPlayer(puregero.multipaper.MultiPaperEntitiesHandler.getControllingPassenger(entity)) &&
+                    (puregero.multipaper.MultiPaper.isRealPlayer(puregero.multipaper.MultiPaperEntitiesHandler.getControllingPassenger(entity)) ||
+                            puregero.multipaper.MultiPaper.isChunkLocal(newChunkHolder))) {
+                puregero.multipaper.MultiPaperEntitiesHandler.onEntityUpdate(entity, packet, newChunkHolder);
+            }
+            // MultiPaper end - Sync entity
             for (ServerPlayerConnection connection : this.seenBy) {
                 if (targetPredicate.test(connection.getPlayer())) {
+                    if (puregero.multipaper.MultiPaper.isExternalPlayer(connection.getPlayer())) continue; // MultiPaper - Sync entity - don't send to external players
                     connection.send(packet);
                 }
             }
diff --git a/net/minecraft/server/level/ServerEntity.java b/net/minecraft/server/level/ServerEntity.java
index 81ed0bc7b685f34101085b986ec391ba70ab29af..d26723ea28268a43431036b1c9185250d73ef289 100644
--- a/net/minecraft/server/level/ServerEntity.java
+++ b/net/minecraft/server/level/ServerEntity.java
@@ -64,7 +64,7 @@ public class ServerEntity {
     private byte lastSentYHeadRot;
     private Vec3 lastSentMovement;
     private int tickCount;
-    private int teleportDelay;
+    public int teleportDelay; // MultiPaper - Sync entity - make public
     private List<Entity> lastPassengers = com.google.common.collect.ImmutableList.of(); // Paper - optimize passenger checks
     private boolean wasRiding;
     private boolean wasOnGround;
@@ -117,6 +117,15 @@ public class ServerEntity {
             this.lastPassengers = passengers;
         }
 
+        // MultiPaper start - Sync entity
+        // in vehicles glitch out when crossing borders, this forces them to sync
+        for (Entity passenger : passengers) {
+            if (passenger.position().distanceToSqr(this.entity.position()) > 16) {
+                passenger.setPosRaw(this.entity.getX(), this.entity.getY(), this.entity.getZ());
+            }
+        }
+        // MultiPaper end - Sync entity
+
         if (!this.trackedPlayers.isEmpty() && this.entity instanceof ItemFrame frame /*&& this.tickCount % 10 == 0*/) { // CraftBukkit - moved tickCount below // Paper - Perf: Only tick item frames if players can see it
             ItemStack itemStack = frame.getItem();
             if (this.level.paperConfig().maps.itemFrameCursorUpdateInterval > 0 && this.tickCount % this.level.paperConfig().maps.itemFrameCursorUpdateInterval == 0 && itemStack.getItem() instanceof MapItem) { // CraftBukkit - Moved this.tickCounter % 10 logic here so item frames do not enter the other blocks // Paper - Make item frame map cursor update interval configurable
@@ -146,7 +155,7 @@ public class ServerEntity {
             byte yRotn = Mth.packDegrees(this.entity.getYRot());
             byte xRotn = Mth.packDegrees(this.entity.getXRot());
             boolean shouldSendRotation = Math.abs(yRotn - this.lastSentYRot) >= 1 || Math.abs(xRotn - this.lastSentXRot) >= 1;
-            if (this.entity.isPassenger()) {
+            if (!(this.entity instanceof ServerPlayer) && this.entity.isPassenger()) { // MultiPaper - Sync entity - force full position update for players
                 if (shouldSendRotation) {
                     this.synchronizer.sendToTrackingPlayers(new ClientboundMoveEntityPacket.Rot(this.entity.getId(), yRotn, xRotn, this.entity.onGround()));
                     this.lastSentYRot = yRotn;
@@ -306,11 +315,13 @@ public class ServerEntity {
     }
 
     public void removePairing(final ServerPlayer player) {
+        if (!puregero.multipaper.MultiPaper.isRealPlayer(player)) return; // MultiPaper - Sync entity - don't send to external players
         this.entity.stopSeenByPlayer(player);
         player.connection.send(new ClientboundRemoveEntitiesPacket(this.entity.getId()));
     }
 
     public void addPairing(final ServerPlayer player) {
+        if (!puregero.multipaper.MultiPaper.isRealPlayer(player)) return; // MultiPaper - Sync entity - don't send to external players
         List<Packet<? super ClientGamePacketListener>> packets = new ArrayList<>();
         this.sendPairingData(player, packets::add);
         player.connection.send(new ClientboundBundlePacket(packets));
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index ba41bdd1d63b9f677c471f4ec5ba9ee3e68672ea..1935652db020f1b1714299590802b074f6b5e804 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -776,7 +776,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         int id = 0;
 
         while (id == 0 || this.chunkSource.hasEntityWithId(id)) {
-            id = ENTITY_COUNTER.incrementAndGet();
+            id = puregero.multipaper.MultiPaperEntityIdManager.next(); // MultiPaper - Sync entity - sync entity ids
         }
 
         return id;
@@ -905,6 +905,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             this.entityTickList
                 .forEach(
                     entity -> {
+                        if (!puregero.multipaper.MultiPaperEntitiesHandler.tickEntity(entity)) return; // MultiPaper - Sync entity - should we tick this entity?
                         if (!entity.isRemoved()) {
                             if (!tickRateManager.isEntityFrozen(entity)) {
                                 profiler.push("checkDespawn");
@@ -1730,6 +1731,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
                 return false;
             }
             // CraftBukkit end
+            puregero.multipaper.MultiPaperEntitiesHandler.onEntitySpawn(entity); // MultiPaper - Sync entity
 
             return this.moonrise$getEntityLookup().addNewEntity(entity); // Paper - rewrite chunk system
         }
@@ -2326,6 +2328,23 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         return entity != null ? entity : this.dragonParts.get(id);
     }
 
+    // MultiPaper start - Sync entity
+    @Nullable
+    public Entity getEntityOrPart(UUID uuid) {
+        Entity entity = this.getEntities().get(uuid);
+        if (entity != null) {
+            return entity;
+        }
+
+        for (Entity part : this.dragonParts.values()) {
+            if (part.getUUID().equals(uuid)) {
+                return part;
+            }
+        }
+        return null;
+    }
+    // MultiPaper end - Sync entity
+
     @Override
     public Collection<EnderDragonPart> dragonParts() {
         return this.dragonParts.values();
diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 46cc56bb08dd709af5bf521d1d3b822b9f83d882..248c43b6825bc0d9e5ed9e4d8ee356fceb6a93f4 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -608,7 +608,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
             // CraftBukkit end
             ValueOutput vehicleWrapper = playerOutput.child("RootVehicle");
             vehicleWrapper.store("Attach", UUIDUtil.CODEC, vehicle.getUUID());
+            rootVehicle.isSavingVehicle = true; // MultiPaper - Sync entity
             rootVehicle.save(vehicleWrapper.child("Entity"));
+            rootVehicle.isSavingVehicle = false; // MultiPaper - Sync entity
         }
     }
 
@@ -1676,7 +1678,7 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
     @Override
     public void take(final Entity entity, final int orgCount) {
         super.take(entity, orgCount);
-        this.containerMenu.broadcastChanges();
+        if (!puregero.multipaper.MultiPaperEntitiesHandler.takingItem) this.containerMenu.broadcastChanges(); // MultiPaper - Sync entity
     }
 
     // CraftBukkit start - moved bed result checks from below into separate method
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 2ccc15e2294b1b0475ae0574c637544279c28095..2f9d6b42506b76d62cceb64aa96776dd6fc1c9dd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -317,6 +317,10 @@ public abstract class Entity
     public static final int FLAG_INVISIBLE = 5;
     protected static final int FLAG_GLOWING = 6;
     protected static final int FLAG_FALL_FLYING = 7;
+    // MultiPaper start - Sync entity
+    public boolean isSyncing = false; // add isSyncing
+    public boolean isSavingVehicle = false; // add isSavingVehicle
+    // MultiPaper end - Sync entity
     private static final EntityDataAccessor<Integer> DATA_AIR_SUPPLY_ID = SynchedEntityData.defineId(Entity.class, EntityDataSerializers.INT);
     private static final EntityDataAccessor<Optional<Component>> DATA_CUSTOM_NAME = SynchedEntityData.defineId(
         Entity.class, EntityDataSerializers.OPTIONAL_COMPONENT
@@ -390,6 +394,7 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public volatile boolean fake = false; // MultiPaper - Sync entity - fake entity nms
 
     public void inactiveTick() {
     }
@@ -757,6 +762,11 @@ public abstract class Entity
     }
     // Paper end
 
+    // MultiPaper start - Sync entity
+    // The Vanilla server never changes an entity's id from the randomly generated one, so we can use Java's built-in equals and hashcode instead
+    // Since MultiPaper does change the entity's id, these can return true when the entities are not equal
+    // The Vanilla client does actually use these methods, but the server does not
+    /*
     @Override
     public boolean equals(final Object obj) {
         return obj instanceof Entity entity && entity.getId() == this.getId();
@@ -766,6 +776,8 @@ public abstract class Entity
     public int hashCode() {
         return this.getId();
     }
+    */
+    // MultiPaper end - Sync entity
 
     public void remove(final Entity.RemovalReason reason) {
         // CraftBukkit start - add Bukkit remove cause
@@ -2163,6 +2175,91 @@ public abstract class Entity
     }
 
     public boolean isInWater() {
+        // MultiPaper start - Sync entity
+        // manually check if touching water for external entities
+        if (puregero.multipaper.MultiPaper.isExternalPlayer(this) || (!puregero.multipaper.MultiPaper.isRealPlayer(this) && !puregero.multipaper.MultiPaper.isChunkLocal(this.level.getChunkIfLoaded(this.blockPosition())))) {
+            // based off CollisionUtil.getCollisionsForBlocksOrWorldBorder
+            final AABB aabb = this.getBoundingBox().deflate(0.001D);
+
+            final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(this.level);
+            final int maxSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMaxSection(this.level);
+
+            final int minBlockX = Mth.floor(aabb.minX);
+            final int maxBlockX = Mth.ceil(aabb.maxX);
+
+            final int minBlockY = Math.max((minSection << 4) - 1, Mth.floor(aabb.minY) - 1);
+            final int maxBlockY = Math.min((maxSection << 4) + 16, Mth.ceil(aabb.maxY) + 1);
+
+            final int minBlockZ = Mth.floor(aabb.minZ);
+            final int maxBlockZ = Mth.ceil(aabb.maxZ);
+
+            // special cases:
+            if (minBlockY > maxBlockY) {
+                // no point in checking
+                return false;
+            }
+
+            final int minChunkX = minBlockX >> 4;
+            final int maxChunkX = maxBlockX >> 4;
+
+            final int minChunkY = minBlockY >> 4;
+            final int maxChunkY = maxBlockY >> 4;
+
+            final int minChunkZ = minBlockZ >> 4;
+            final int maxChunkZ = maxBlockZ >> 4;
+
+            for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
+                for (int currChunkX = minChunkX; currChunkX <= maxChunkX; ++currChunkX) {
+                    final net.minecraft.world.level.chunk.ChunkAccess chunk = this.level.getChunkIfLoadedImmediately(currChunkX, currChunkZ);
+
+                    if (chunk == null) {
+                        return false; // if we're touching an unloaded chunk then it's false
+                    }
+
+                    final net.minecraft.world.level.chunk.LevelChunkSection[] sections = chunk.getSections();
+
+                    // bound y
+                    for (int currChunkY = minChunkY; currChunkY <= maxChunkY; ++currChunkY) {
+                        final int sectionIdx = currChunkY - minSection;
+                        if (sectionIdx < 0 || sectionIdx >= sections.length) {
+                            continue;
+                        }
+
+                        final net.minecraft.world.level.chunk.LevelChunkSection section = sections[sectionIdx];
+                        if (section == null || section.hasOnlyAir()) {
+                            continue;
+                        }
+
+                        final int sectionAdjust = section.moonrise$hasSpecialCollidingBlocks() ? 0 : 1;
+                        final net.minecraft.world.level.chunk.PalettedContainer<BlockState> blocks = section.getStates();
+
+                        final int minXIterate = currChunkX == minChunkX ? (minBlockX & 15) + sectionAdjust : 0;
+                        final int maxXIterate = currChunkX == maxChunkX ? (maxBlockX & 15) - sectionAdjust : 15;
+                        final int minZIterate = currChunkZ == minChunkZ ? (minBlockZ & 15) + sectionAdjust : 0;
+                        final int maxZIterate = currChunkZ == maxChunkZ ? (maxBlockZ & 15) - sectionAdjust : 15;
+                        final int minYIterate = currChunkY == minChunkY ? (minBlockY & 15) + sectionAdjust : 0;
+                        final int maxYIterate = currChunkY == maxChunkY ? (maxBlockY & 15) - sectionAdjust : 15;
+
+                        for (int currY = minYIterate; currY <= maxYIterate; ++currY) {
+                            for (int currZ = minZIterate; currZ <= maxZIterate; ++currZ) {
+                                for (int currX = minXIterate; currX <= maxXIterate; ++currX) {
+                                    final int localBlockIndex = (currX) | (currZ << 4) | ((currY) << 8);
+                                    final FluidState fluid = blocks.get(localBlockIndex).getFluidState();
+
+                                    if (fluid.is(FluidTags.WATER)) {
+                                        return true;
+                                    }
+                                }
+                            }
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+        // MultiPaper end - Sync entity
+
         return this.wasTouchingWater;
     }
 
@@ -2665,7 +2762,7 @@ public abstract class Entity
         }
 
         String id = this.getEncodeId(includeNonSaveable); // Paper - Raw entity serialization API
-        if ((!this.persist && !forceSerialization) || id == null) { // CraftBukkit - persist flag // Paper - Raw entity serialization API
+        if ((!this.isSyncing && !this.persist && !forceSerialization) || id == null) { // CraftBukkit - persist flag // Paper - Raw entity serialization API // MultiPaper - Sync entity - only use the persist flag if we're not syncing between servers
             return false;
         }
 
@@ -2675,7 +2772,7 @@ public abstract class Entity
     }
 
     public boolean save(final ValueOutput output) {
-        return !this.isPassenger() && this.saveAsPassenger(output);
+        return (this.isSyncing || !this.isPassenger()) && this.saveAsPassenger(output); // MultiPaper - Sync entity - save passengers when syncing
     }
 
     public void saveWithoutId(final ValueOutput output) {
@@ -2688,6 +2785,15 @@ public abstract class Entity
         // Paper end - Raw entity serialization API
         // CraftBukkit end
         try {
+            // MultiPaper start - Sync entity
+            if (this.isSyncing && puregero.multipaper.config.MultiPaperConfiguration.get().syncSettings.syncEntityIds) {
+                output.putInt("EntityId", this.id);
+            }
+            if (this.isSavingVehicle && puregero.multipaper.config.MultiPaperConfiguration.get().syncSettings.persistentVehicleEntityIdsSeconds > 0) {
+                output.putInt("VehicleEntityId", this.id);
+                output.putLong("VehicleEntityIdExpireTime", System.currentTimeMillis());
+            }
+            // MultiPaper end - Sync entity
             if (includeAll) { // CraftBukkit - selectively save position
             if (this.vehicle != null) {
                 output.store("Pos", Vec3.CODEC, new Vec3(this.vehicle.getX(), this.getY(), this.vehicle.getZ()));
@@ -2781,10 +2887,18 @@ public abstract class Entity
                 ValueOutput.ValueOutputList passengersList = output.childrenList("Passengers");
 
                 for (Entity passenger : this.getPassengers()) {
+                    // MultiPaper start - Sync entity
+                    if (this.isSyncing) passenger.isSyncing = true;
+                    if (this.isSavingVehicle) passenger.isSavingVehicle = true;
+                    // MultiPaper end - Sync entity
                     ValueOutput passengerOutput = passengersList.addChild();
                     if (!passenger.saveAsPassenger(passengerOutput, includeAll, includeNonSaveable, forceSerialization)) { // CraftBukkit - pass on includeAll // Paper - Raw entity serialization API
                         passengersList.discardLast();
                     }
+                    // MultiPaper start - Sync entity
+                    if (this.isSyncing) passenger.isSyncing = false;
+                    if (this.isSavingVehicle) passenger.isSavingVehicle = false;
+                    // MultiPaper end - Sync entity
                 }
 
                 if (passengersList.isEmpty()) {
@@ -2833,6 +2947,36 @@ public abstract class Entity
 
     public void load(final ValueInput input) {
         try {
+            // MultiPaper start - Sync entity
+            if (!(this instanceof ServerPlayer)) {
+                if (puregero.multipaper.config.MultiPaperConfiguration.get().syncSettings.syncEntityIds) {
+                    input.getInt("EntityId").ifPresent(entityId -> {
+                        if (this.id != entityId) {
+                            if (this.level.getEntity(this.id) == null) {
+                                this.id = entityId;
+                            } else {
+                                LOGGER.error("Failed to sync entity id {} for {}. This entity id is already mapped to: {}", entityId, this, this.level.getEntity(this.id));
+                            }
+                        }
+                    });
+                }
+
+                if (input.getInt("VehicleEntityId").isPresent()) {
+                    int entityId = input.getInt("VehicleEntityId").get();
+                    input.getLong("VehicleEntityIdExpireTime").ifPresent(expireTime -> {
+                        if (expireTime > System.currentTimeMillis() - puregero.multipaper.config.MultiPaperConfiguration.get().syncSettings.persistentVehicleEntityIdsSeconds * 1000L) {
+                            if (this.id != entityId) {
+                                if (this.level.getEntity(this.id) == null) {
+                                    this.id = entityId;
+                                } else {
+                                    LOGGER.error("Failed to persist vehicle entity id {} for {}. This entity id is already mapped to: {}", entityId, this, this.level.getEntity(this.id));
+                                }
+                            }
+                        }
+                    });
+                }
+            }
+            // MultiPaper end - Sync entity
             Vec3 pos = input.read("Pos", Vec3.CODEC).orElse(Vec3.ZERO);
             Vec3 motion = input.read("Motion", Vec3.CODEC).orElse(Vec3.ZERO);
             Vec2 rotation = input.read("Rotation", Vec2.CODEC).orElse(Vec2.ZERO);
@@ -2857,6 +3001,7 @@ public abstract class Entity
                 this.uuid = id;
                 this.stringUUID = this.uuid.toString();
             });
+            if (this instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragon enderDragon) enderDragon.regenerateUUIDs(); // MultiPaper - Sync entity
             if (!Double.isFinite(this.getX()) || !Double.isFinite(this.getY()) || !Double.isFinite(this.getZ())) {
                 throw new IllegalStateException("Entity has invalid position");
             }
@@ -2963,7 +3108,7 @@ public abstract class Entity
     }
 
     public final @Nullable String getEncodeId(final boolean includeNonSaveable) {
-        if (!includeNonSaveable && !this.getType().canSerialize()) {
+        if (!includeNonSaveable && !this.getType().canSerialize() && (!this.isSyncing || this instanceof ServerPlayer)) { // MultiPaper - Sync entity
             // Paper end - Raw entity serialization API
             return null;
         }
@@ -3401,7 +3546,7 @@ public abstract class Entity
             if (entityToRide.getBukkitEntity() instanceof org.bukkit.entity.Vehicle && this.getBukkitEntity() instanceof org.bukkit.entity.LivingEntity) {
                 org.bukkit.event.vehicle.VehicleEnterEvent event = new org.bukkit.event.vehicle.VehicleEnterEvent((org.bukkit.entity.Vehicle) entityToRide.getBukkitEntity(), this.getBukkitEntity());
                 // Suppress during worldgen
-                if (this.valid) {
+                if (this.valid && !puregero.multipaper.MultiPaperEntitiesHandler.modifyingPassengers) { // MultiPaper - Sync entity - don't call VehicleEnterEvent if we're syncing the passengers from another server
                     org.bukkit.Bukkit.getPluginManager().callEvent(event);
                 }
                 if (event.isCancelled()) {
@@ -3411,7 +3556,7 @@ public abstract class Entity
 
             org.bukkit.event.entity.EntityMountEvent event = new org.bukkit.event.entity.EntityMountEvent(this.getBukkitEntity(), entityToRide.getBukkitEntity());
             // Suppress during worldgen
-            if (this.valid) {
+            if (this.valid && !puregero.multipaper.MultiPaperEntitiesHandler.modifyingPassengers) { // MultiPaper - Sync entity - don't call EntityMountEvent if we're syncing the passengers from another server
                 org.bukkit.Bukkit.getPluginManager().callEvent(event);
             }
             if (event.isCancelled()) {
@@ -3456,6 +3601,16 @@ public abstract class Entity
         }
     }
 
+    // MultiPaper start - Sync entity
+    // allow for removing passengers when unloading an external chunk
+    public void forceEjectPassengers() {
+        for (Entity passenger : this.passengers) {
+            passenger.vehicle = null;
+        }
+        this.passengers = ImmutableList.of();
+    }
+    // MultiPaper end - Sync entity
+
     public void removeVehicle() {
         // Paper start - Force entity dismount during teleportation
         this.removeVehicle(false);
@@ -3472,6 +3627,7 @@ public abstract class Entity
             if (removalReason == null || removalReason.shouldDestroy()) {
                 this.level().gameEvent(this, GameEvent.ENTITY_DISMOUNT, oldVehicle.position);
             }
+            this.closeInventoriesRecursive(oldVehicle.getRootVehicle()); // MultiPaper - Sync entity - close any open inventories on the vehicle
             // MultiPaper start - Sync player
             // inform other servers that the player is no longer riding the entity and hand them the entity's data
             ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder newChunkHolder = puregero.multipaper.MultiPaper.getChunkHolder(oldVehicle);
@@ -3490,6 +3646,23 @@ public abstract class Entity
         }
     }
 
+    // MultiPaper start - Sync entity
+    // close inventories when an entity changes servers
+    private void closeInventoriesRecursive(Entity entity) {
+        if (entity instanceof net.minecraft.world.Container container) {
+            for (org.bukkit.entity.HumanEntity viewer : new ArrayList<>(container.getViewers())) {
+                if (viewer instanceof org.bukkit.craftbukkit.entity.CraftPlayer craftPlayer) {
+                    craftPlayer.closeInventory(org.bukkit.event.inventory.InventoryCloseEvent.Reason.UNLOADED);
+                }
+            }
+        }
+
+        for (Entity passenger : entity.getPassengers()) {
+            closeInventoriesRecursive(passenger);
+        }
+    }
+    // MultiPaper end - Sync entity
+
     public void stopRiding() {
         // Paper start - Force entity dismount during teleportation
         this.stopRiding(false);
@@ -3545,7 +3718,7 @@ public abstract class Entity
                     (org.bukkit.entity.LivingEntity) passenger.getBukkitEntity(), !suppressCancellation // Paper - Force entity dismount during teleportation
             );
             // Suppress during worldgen
-            if (this.valid) {
+            if (this.valid && !puregero.multipaper.MultiPaperEntitiesHandler.modifyingPassengers) { // MultiPaper - Sync entity - don't call VehicleExitEvent if we're syncing the passengers from another server
                 org.bukkit.Bukkit.getPluginManager().callEvent(event);
             }
             org.bukkit.craftbukkit.entity.CraftEntity craftn = (org.bukkit.craftbukkit.entity.CraftEntity) passenger.getBukkitEntity().getVehicle();
@@ -3557,7 +3730,7 @@ public abstract class Entity
 
         org.bukkit.event.entity.EntityDismountEvent event = new org.bukkit.event.entity.EntityDismountEvent(passenger.getBukkitEntity(), this.getBukkitEntity(), !suppressCancellation); // Paper - Force entity dismount during teleportation
         // Suppress during worldgen
-        if (this.valid) {
+        if (this.valid && !puregero.multipaper.MultiPaperEntitiesHandler.modifyingPassengers) { // MultiPaper - Sync entity - don't call EntityDismountEvent if we're syncing the passengers from another server
             org.bukkit.Bukkit.getPluginManager().callEvent(event);
         }
         if (event.isCancelled()) {
@@ -4196,6 +4369,13 @@ public abstract class Entity
         }
         // Paper end - Fix item duplication and teleport issues
         if (this.level() instanceof ServerLevel serverLevel && !this.isRemoved()) {
+            // MultiPaper start - Sync entity
+            ca.spottedleaf.moonrise.patches.chunk_system.scheduling.NewChunkHolder newChunkHolder = puregero.multipaper.MultiPaper.getChunkHolder(this);
+            if (newChunkHolder != null && newChunkHolder.externalOwner != null && !newChunkHolder.externalOwner.isMe()) {
+                newChunkHolder.externalOwner.getConnection().send(new puregero.multipaper.externalserverprotocol.EntityTeleportPacket(this, transition));
+            }
+            // MultiPaper end - Sync entity
+
             // CraftBukkit start
             PositionMoveRotation absolutePosition = PositionMoveRotation.calculateAbsolute(PositionMoveRotation.of(this), PositionMoveRotation.of(transition), transition.relatives());
             Vec3 velocity = absolutePosition.deltaMovement(); // Paper
@@ -4796,7 +4976,7 @@ public abstract class Entity
         return false;
     }
 
-    private Stream<Entity> getIndirectPassengersStream() {
+    public Stream<Entity> getIndirectPassengersStream() { // MultiPaper - Sync entity - Make public
         if (this.passengers.isEmpty()) { return Stream.of(); } // Paper - Optimize indirect passenger iteration
         return this.passengers.stream().flatMap(Entity::getSelfAndPassengers);
     }
@@ -5057,6 +5237,7 @@ public abstract class Entity
     public void addDeltaMovement(final Vec3 momentum) {
         if (momentum.isFinite()) {
             this.setDeltaMovement(this.getDeltaMovement().add(momentum));
+            puregero.multipaper.externalserverprotocol.AddDeltaMovementPacket.broadcast(this, momentum); // MultiPaper - Sync entity
         }
     }
 
@@ -5161,6 +5342,13 @@ public abstract class Entity
             int fx = Mth.floor(x);
             int fy = Mth.floor(y);
             int fz = Mth.floor(z);
+
+            // MultiPaper start - Sync entity
+            if ((fx >> 4) != (this.blockPosition.getX() >> 4) || (fz >> 4) != (this.blockPosition.getZ() >> 4)) {
+                puregero.multipaper.MultiPaperEntitiesHandler.onChunkMove(this, new BlockPos(fx, fy, fz), this.blockPosition);
+            }
+            // MultiPaper end - Sync entity
+
             if (fx != this.blockPosition.getX() || fy != this.blockPosition.getY() || fz != this.blockPosition.getZ()) {
                 this.blockPosition = new BlockPos(fx, fy, fz);
                 this.inBlockState = null;
@@ -5307,6 +5495,12 @@ public abstract class Entity
         // Paper end - rewrite chunk system
         org.bukkit.craftbukkit.event.CraftEventFactory.callEntityRemoveEvent(this, cause); // CraftBukkit
         final boolean alreadyRemoved = this.removalReason != null; // Paper - Folia schedulers
+
+        // MultiPaper start - Sync entity
+        puregero.multipaper.MultiPaperEntitiesHandler.onEntityRemove(this, reason);
+        puregero.multipaper.externalserverprotocol.EntityRemovePacket.setEntityRemoved(this.getUUID(), reason);
+        // MultiPaper end - Sync entity
+
         if (this.removalReason == null) {
             this.removalReason = reason;
         }
@@ -5352,7 +5546,7 @@ public abstract class Entity
 
     @Override
     public boolean shouldBeSaved() {
-        return (this.removalReason == null || this.removalReason.shouldSave())
+        return this.type.canSerialize() && (this.removalReason == null || this.removalReason.shouldSave()) // MultiPaper - Sync entity - if they can't be serialized, they also can't be saved
             && !this.isPassenger()
             && (!this.isVehicle() || !((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this).moonrise$hasAnyPlayerPassengers()); // Paper - rewrite chunk system
     }
@@ -5612,4 +5806,19 @@ public abstract class Entity
         return false;
     }
     // Purpur end - Ridables
+
+    // MultiPaper start - Sync entity
+    // fake entity nms
+    public boolean isFake() {
+        return this.fake;
+    }
+
+    public void setFake() {
+        // we need to make sure that actual players are not set fake
+        if (this instanceof ServerPlayer player && player.server.getPlayerList().getPlayers().contains(player)) {
+            throw new IllegalStateException("Online player cannot be set as fake entity");
+        }
+        this.fake = true;
+    }
+    // MultiPaper end - Sync entity
 }
diff --git a/net/minecraft/world/entity/Leashable.java b/net/minecraft/world/entity/Leashable.java
index 98034a4c96bf2972316078d3b3a49140921aab50..c2156153ea69c5d22f5da046053ed35b342cfff1 100644
--- a/net/minecraft/world/entity/Leashable.java
+++ b/net/minecraft/world/entity/Leashable.java
@@ -88,7 +88,7 @@ public interface Leashable {
         output.storeNullable("leash", Leashable.LeashData.CODEC, leashData);
     }
 
-    private static <E extends Entity & Leashable> void restoreLeashFromSave(final E entity, final Leashable.LeashData leashData) {
+    public static <E extends Entity & Leashable> void restoreLeashFromSave(final E entity, final Leashable.LeashData leashData) { // MultiPaper - Sync entity - make public
         if (leashData.delayedLeashInfo != null && entity.level() instanceof ServerLevel serverLevel) {
             Optional<UUID> leashUuid = leashData.delayedLeashInfo.left();
             Optional<BlockPos> pos = leashData.delayedLeashInfo.right();
@@ -127,7 +127,7 @@ public interface Leashable {
             entity.setLeashData(null);
             entity.onLeashRemoved();
             if (entity.level() instanceof ServerLevel level) {
-                if (dropLead) {
+                if (dropLead && puregero.multipaper.MultiPaper.isChunkLocal(entity.level().getChunkIfLoaded(entity.blockPosition()))) { // MultiPaper - Sync entity - only drop leads on the local server
                     entity.spawnAtLocation(level, Items.LEAD);
                 }
 
diff --git a/net/minecraft/world/entity/LivingEntity.java b/net/minecraft/world/entity/LivingEntity.java
index 2941842157754d71163632573e3eab7a9d9b496b..646ed86e7ff0d739d7c0a962e6f814eab161b745 100644
--- a/net/minecraft/world/entity/LivingEntity.java
+++ b/net/minecraft/world/entity/LivingEntity.java
@@ -611,6 +611,14 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
         return true;
     }
 
+    // MultiPaper start - Sync entity
+    public void tickDeathIfDead() {
+        if (this.isDeadOrDying() && this.level().shouldTickDeath(this)) {
+            this.tickDeath();
+        }
+    }
+    // MultiPaper end - Sync entity
+
     protected void tickDeath() {
         this.deathTime++;
         if (this.deathTime >= 20 && !this.level().isClientSide() && !this.isRemoved()) {
@@ -1311,11 +1319,12 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     protected void onEffectAdded(final MobEffectInstance effect, final @Nullable Entity source) {
-        if (!this.level().isClientSide()) {
+        if (!this.level().isClientSide() && !puregero.multipaper.MultiPaper.isExternalPlayer(this)) { // MultiPaper - Sync entity
             this.effectsDirty = true;
             effect.getEffect().value().addAttributeModifiers(this.getAttributes(), effect.getAmplifier());
             this.sendEffectToPassengers(effect);
         }
+        puregero.multipaper.MultiPaperEffectsHandler.onEffectAdd(this, effect); // MultiPaper - Sync entity
     }
 
     public void sendEffectToPassengers(final MobEffectInstance effect) {
@@ -1327,7 +1336,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
     }
 
     protected void onEffectUpdated(final MobEffectInstance effect, final boolean doRefreshAttributes, final @Nullable Entity source) {
-        if (!this.level().isClientSide()) {
+        if (!this.level().isClientSide() && !puregero.multipaper.MultiPaper.isExternalPlayer(this)) { // MultiPaper - Sync entity
             this.effectsDirty = true;
             if (doRefreshAttributes) {
                 MobEffect mobEffect = effect.getEffect().value();
@@ -1338,10 +1347,11 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
             this.sendEffectToPassengers(effect);
         }
+        puregero.multipaper.MultiPaperEffectsHandler.onEffectAdd(this, effect); // MultiPaper - Sync entity
     }
 
     protected void onEffectsRemoved(final Collection<MobEffectInstance> effects) {
-        if (!this.level().isClientSide()) {
+        if (!this.level().isClientSide() && !puregero.multipaper.MultiPaper.isExternalPlayer(this)) { // MultiPaper - Sync entity
             this.effectsDirty = true;
 
             for (MobEffectInstance effect : effects) {
@@ -1356,6 +1366,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
             this.refreshDirtyAttributes();
         }
+        effects.forEach(effect -> puregero.multipaper.MultiPaperEffectsHandler.onEffectRemove(this, effect)); // MultiPaper - Sync entity
     }
 
     private void refreshDirtyAttributes() {
@@ -1457,6 +1468,13 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
 
     @Override
     public boolean hurtServer(final ServerLevel level, final DamageSource source, float damage) {
+        // MultiPaper start - Sync entity
+        puregero.multipaper.ExternalServerConnection owner = puregero.multipaper.MultiPaperEntityInteractHandler.getOwner(this);
+        if (owner != null) {
+            owner.send(new puregero.multipaper.externalserverprotocol.HurtEntityPacket(this, source, damage));
+            return false;
+        }
+        // MultiPaper end - Sync entity
         if (this.isInvulnerableTo(level, source)) {
             return false;
         }
@@ -3669,7 +3687,7 @@ public abstract class LivingEntity extends Entity implements Attackable, Waypoin
                 // Paper start - EntityEquipmentChangedEvent, PlayerArmorChangeEvent
                 final org.bukkit.inventory.ItemStack oldItem = CraftItemStack.asBukkitCopy(previous);
                 final org.bukkit.inventory.ItemStack newItem = CraftItemStack.asBukkitCopy(current);
-                if (this instanceof ServerPlayer && slot.getType() == EquipmentSlot.Type.HUMANOID_ARMOR) {
+                if (this instanceof ServerPlayer && puregero.multipaper.MultiPaper.isRealPlayer(this) && slot.getType() == EquipmentSlot.Type.HUMANOID_ARMOR) { // MultiPaper - Sync entity - only run PlayerArmorChangeEvent for local players
                     new com.destroystokyo.paper.event.player.PlayerArmorChangeEvent((org.bukkit.entity.Player)this.getBukkitEntity(), com.destroystokyo.paper.event.player.PlayerArmorChangeEvent.SlotType.valueOf(slot.name()), oldItem, newItem).callEvent();
                 }
                 // Paper end - EntityEquipmentChangedEvent, PlayerArmorChangeEvent
diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 862224dfc003bf02ee0d7c1d804549210d6c8ffa..c119f1a953ca690128bf48a048f2a41cdd237170 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -37,7 +37,7 @@ public abstract class PathNavigation {
     protected final Mob mob;
     protected final Level level;
     protected @Nullable Path path;
-    protected double speedModifier;
+    public double speedModifier; // MultiPaper - Sync entity - make public
     protected int tick;
     protected int lastStuckCheck;
     protected Vec3 lastStuckCheckPos = Vec3.ZERO;
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index 773351cd6c40e57a9c81d3d1b25161f282ad09c2..47a75a88f3c7df92bd08ae137543cc5b98de71b3 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -96,14 +96,17 @@ public class EnderDragon extends Mob implements Enemy {
 
     public EnderDragon(final EntityType<? extends EnderDragon> type, final Level level) {
         super(EntityTypes.ENDER_DRAGON, level);
-        this.head = new EnderDragonPart(this, "head", 1.0F, 1.0F);
-        this.neck = new EnderDragonPart(this, "neck", 3.0F, 3.0F);
-        this.body = new EnderDragonPart(this, "body", 5.0F, 3.0F);
-        this.tail1 = new EnderDragonPart(this, "tail", 2.0F, 2.0F);
-        this.tail2 = new EnderDragonPart(this, "tail", 2.0F, 2.0F);
-        this.tail3 = new EnderDragonPart(this, "tail", 2.0F, 2.0F);
-        this.wing1 = new EnderDragonPart(this, "wing", 4.0F, 2.0F);
-        this.wing2 = new EnderDragonPart(this, "wing", 4.0F, 2.0F);
+        // MultiPaper start - Sync entity
+        // adds unique identifier to each ender dragon part to make them uniquely identifiable
+        this.head = new EnderDragonPart(this, "head", 1.0F, 1.0F, "head");
+        this.neck = new EnderDragonPart(this, "neck", 3.0F, 3.0F, "neck");
+        this.body = new EnderDragonPart(this, "body", 5.0F, 3.0F, "body");
+        this.tail1 = new EnderDragonPart(this, "tail", 2.0F, 2.0F, "tail1");
+        this.tail2 = new EnderDragonPart(this, "tail", 2.0F, 2.0F, "tail2");
+        this.tail3 = new EnderDragonPart(this, "tail", 2.0F, 2.0F, "tail3");
+        this.wing1 = new EnderDragonPart(this, "wing", 4.0F, 2.0F, "wing1");
+        this.wing2 = new EnderDragonPart(this, "wing", 4.0F, 2.0F, "wing2");
+        // MultiPaper end - Sync entity
         this.subEntities = new EnderDragonPart[]{this.head, this.neck, this.body, this.tail1, this.tail2, this.tail3, this.wing1, this.wing2};
         this.setHealth(this.getMaxHealth());
         this.noPhysics = true;
@@ -129,8 +132,26 @@ public class EnderDragon extends Mob implements Enemy {
             }
         };
         // Purpur end - Ridables
+        // MultiPaper start - Sync entity
+        int DRAGON_ENTITY_COUNT = 9;
+        this.setId(puregero.multipaper.MultiPaperEntityIdManager.next(DRAGON_ENTITY_COUNT)[0]);
+        this.regenerateUUIDs();
+        // MultiPaper end - Sync entity
     }
 
+    // MultiPaper start
+    public void regenerateUUIDs() {
+        this.head.regenerateUUID(this, "head", this.getId() + 1);
+        this.neck.regenerateUUID(this, "neck", this.getId() + 2);
+        this.body.regenerateUUID(this, "body", this.getId() + 3);
+        this.tail1.regenerateUUID(this, "tail1", this.getId() + 4);
+        this.tail2.regenerateUUID(this, "tail2", this.getId() + 5);
+        this.tail3.regenerateUUID(this, "tail3", this.getId() + 6);
+        this.wing1.regenerateUUID(this, "wing1", this.getId() + 7);
+        this.wing2.regenerateUUID(this, "wing2", this.getId() + 8);
+    }
+    // MultiPaper end
+
     // Purpur start - Ridables
     @Override
     public boolean isRidable() {
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragonPart.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragonPart.java
index a677aed54e16f3150ea1eabe22935a3d0d1507c4..713f13ff91def55e704e736c3164034c36b5a764 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragonPart.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragonPart.java
@@ -19,14 +19,41 @@ public class EnderDragonPart extends Entity {
     public final String name;
     private final EntityDimensions size;
 
-    public EnderDragonPart(final EnderDragon parentMob, final String name, final float w, final float h) {
+    public EnderDragonPart(final EnderDragon parentMob, final String name, final float w, final float h, final String uuidSuffix) { // MultiPaper - Sync entity
         super(parentMob.getType(), parentMob.level());
         this.size = EntityDimensions.scalable(w, h);
         this.refreshDimensions();
         this.parentMob = parentMob;
         this.name = name;
+        // MultiPaper start - Sync entity
+        this.uuid = this.deterministicUUID(parentMob.getUUID().toString(), uuidSuffix);
+        this.stringUUID = this.uuid.toString();
+        // MultiPaper end - Sync entity
     }
 
+    // MultiPaper start - Sync entity
+    public void regenerateUUID(EnderDragon owner, String uuidSuffix, int id) {
+        this.setId(id);
+        this.uuid = this.deterministicUUID(owner.getUUID().toString(), uuidSuffix);
+        this.stringUUID = this.uuid.toString();
+    }
+
+    private java.util.UUID deterministicUUID(String str1, String str2) {
+        try {
+            java.security.MessageDigest digest = java.security.MessageDigest.getInstance("SHA-256");
+            byte[] hash1 = digest.digest(str1.getBytes(java.nio.charset.StandardCharsets.UTF_8));
+            byte[] hash2 = digest.digest(str2.getBytes(java.nio.charset.StandardCharsets.UTF_8));
+            byte[] combinedHash = new byte[hash1.length + hash2.length];
+            System.arraycopy(hash1, 0, combinedHash, 0, hash1.length);
+            System.arraycopy(hash2, 0, combinedHash, hash1.length, hash2.length);
+
+            return java.util.UUID.nameUUIDFromBytes(combinedHash);
+        } catch (java.security.NoSuchAlgorithmException e) {
+            throw new RuntimeException(e);
+        }
+    }
+    // MultiPaper end - Sync entity
+
     // Purpur start - Ridables
     @Override
     public net.minecraft.world.InteractionResult interact(final net.minecraft.world.entity.player.Player player, final net.minecraft.world.InteractionHand hand, final net.minecraft.world.phys.Vec3 location) {
diff --git a/net/minecraft/world/entity/npc/villager/AbstractVillager.java b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
index 8c55eb2cc6f48643d29b50a12a44be19fd29b4bc..2f629f0369aeb7dac53d4d7db1645e6d17f05bbb 100644
--- a/net/minecraft/world/entity/npc/villager/AbstractVillager.java
+++ b/net/minecraft/world/entity/npc/villager/AbstractVillager.java
@@ -128,6 +128,7 @@ public abstract class AbstractVillager extends AgeableMob implements Npc, Mercha
     public MerchantOffers getOffers() {
         if (this.level() instanceof ServerLevel serverLevel) {
             if (this.offers == null) {
+                if (!io.papermc.paper.util.MCUtil.isMainThread()) return new MerchantOffers(); // MultiPaper - Sync entity - don't calculate trade offers async
                 this.offers = new MerchantOffers();
                 this.updateTrades(serverLevel);
             }
diff --git a/net/minecraft/world/entity/projectile/FishingHook.java b/net/minecraft/world/entity/projectile/FishingHook.java
index c0bb1b62dfcdebbd34a8e8423fa36eb4924a55e5..ed1f2ec2777bcbca34b3ef06c5f27260310cbf37 100644
--- a/net/minecraft/world/entity/projectile/FishingHook.java
+++ b/net/minecraft/world/entity/projectile/FishingHook.java
@@ -315,6 +315,7 @@ public class FishingHook extends Projectile {
     }
 
     private void catchingFish(final BlockPos blockPos) {
+        if (!puregero.multipaper.MultiPaper.isChunkLocal(this.level().getChunkIfLoaded(blockPosition()))) return; // MultiPaper - Sync entity - only catch fish on our own chunks
         ServerLevel serverLevel = (ServerLevel)this.level();
         int fishingSpeed = 1;
         BlockPos above = blockPos.above();
diff --git a/net/minecraft/world/entity/projectile/Projectile.java b/net/minecraft/world/entity/projectile/Projectile.java
index 7f3673a44b0200d18dbf7f1fac8d377bfc994232..fd9fd2ad3ca932b93ce236154ef16d7dadb589d5 100644
--- a/net/minecraft/world/entity/projectile/Projectile.java
+++ b/net/minecraft/world/entity/projectile/Projectile.java
@@ -45,6 +45,7 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     public boolean hasBeenShot = false;
     private @Nullable Entity lastDeflectedBy;
     protected boolean hitCancelled = false; // CraftBukkit
+    private boolean hasHitEntity; // MultiPaper - Sync entity
 
     protected Projectile(final EntityType<? extends Projectile> type, final Level level) {
         super(type, level);
@@ -373,9 +374,28 @@ public abstract class Projectile extends Entity implements TraceableEntity {
     protected void onItemBreak(final Item item) {
     }
 
+    // MultiPaper start - Sync entity
+    // expose onHit
+    public void onHit0(final HitResult hitResult) {
+        this.onHit(hitResult);
+    }
+    // MultiPaper end - Sync entity
+
     protected void onHit(final HitResult hitResult) {
         HitResult.Type type = hitResult.getType();
         if (type == HitResult.Type.ENTITY) {
+            // MultiPaper start - Sync entity
+            // handle entity hits on their server
+            if (this.hasHitEntity) {
+                return; // prevent infinite hits of the entity if an external server ignores the projectile hit packet
+            }
+            hasHitEntity = true;
+            puregero.multipaper.ExternalServerConnection entityOwner = puregero.multipaper.MultiPaperEntityInteractHandler.getOwner(((EntityHitResult) hitResult).getEntity());
+            if (entityOwner != null) {
+                entityOwner.send(new puregero.multipaper.externalserverprotocol.ProjectileHitEntityPacket(this, ((EntityHitResult) hitResult).getEntity(), hitResult.getLocation()));
+                return;
+            }
+            // MultiPaper end - Sync entity
             EntityHitResult entityHitResult = (EntityHitResult)hitResult;
             Entity entityHit = entityHitResult.getEntity();
             if (entityHit.is(EntityTypeTags.REDIRECTABLE_PROJECTILE) && entityHit instanceof Projectile projectile) {
diff --git a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
index 578eefd44f0eaba4f2de3cfe58dedb70b9bcc60e..98a88ed3d4fa1317d4184468dc2e4962d0b95fc4 100644
--- a/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
+++ b/net/minecraft/world/entity/projectile/throwableitemprojectile/ThrownEnderpearl.java
@@ -98,7 +98,7 @@ public class ThrownEnderpearl extends ThrowableItemProjectile {
                 );
         }
 
-        if (this.level() instanceof ServerLevel level && !this.isRemoved()) {
+        if (this.level() instanceof ServerLevel level && !this.isRemoved() && puregero.multipaper.MultiPaper.isChunkLocal((ServerLevel) this.level(), this.blockPosition())) { // MultiPaper - Sync entity - fix ender pearls on external entities
             Entity owner = this.getOwner();
             if (owner != null && isAllowedToTeleportOwner(owner, level)) {
                 Vec3 teleportPos = this.oldPosition();
diff --git a/net/minecraft/world/entity/raid/Raid.java b/net/minecraft/world/entity/raid/Raid.java
index 16d789ffce1f645505f51a1e61944cd2bc0672c5..f86f597acf955c8b69c08e8e8bc54fb20bf0c59a 100644
--- a/net/minecraft/world/entity/raid/Raid.java
+++ b/net/minecraft/world/entity/raid/Raid.java
@@ -73,6 +73,10 @@ public class Raid {
     public static final SpawnPlacementType RAVAGER_SPAWN_PLACEMENT_TYPE = SpawnPlacements.getPlacementType(EntityTypes.RAVAGER);
     public static final MapCodec<Raid> MAP_CODEC = RecordCodecBuilder.mapCodec(
         i -> i.group(
+                // MultiPaper start - Sync entity
+                Codec.INT.fieldOf("id").forGetter(raid -> raid.id),
+                UUIDUtil.CODEC.fieldOf("level").forGetter(raid -> raid.level),
+                // MultiPaper end - Sync entity
                 Codec.BOOL.fieldOf("started").forGetter(r -> r.started),
                 Codec.BOOL.fieldOf("active").forGetter(r -> r.active),
                 Codec.LONG.fieldOf("ticks_active").forGetter(r -> r.ticksActive),
@@ -115,6 +119,10 @@ public class Raid {
     private final Map<Integer, Raider> groupToLeaderMap = Maps.newHashMap();
     private final Map<Integer, Set<Raider>> groupRaiderMap = Maps.newHashMap();
     public final Set<UUID> heroesOfTheVillage = Sets.newHashSet();
+    // MultiPaper start - Sync entity
+    public int id;
+    public UUID level;
+    // MultiPaper end - Sync entity
     public long ticksActive;
     private BlockPos center;
     private boolean started;
@@ -133,7 +141,11 @@ public class Raid {
     private int celebrationTicks;
     private Optional<BlockPos> waveSpawnPos = Optional.empty();
 
-    public Raid(final BlockPos center, final Difficulty difficulty) {
+    // MultiPaper start - Sync entity
+    public Raid(final int id, final UUID level, final BlockPos center, final Difficulty difficulty) {
+        this.id = id;
+        this.level = level;
+    // MultiPaper end - Sync entity
         this.active = true;
         this.raidCooldownTicks = 300;
         this.raidEvent.setProgress(0.0F);
@@ -144,6 +156,10 @@ public class Raid {
     }
 
     private Raid(
+        // MultiPaper start - Sync entity
+        final int id,
+        final UUID level,
+        // MultiPaper end - Sync entity
         final boolean started,
         final boolean active,
         final long ticksActive,
@@ -158,8 +174,12 @@ public class Raid {
         final Set<UUID> heroesOfTheVillage
         , final Optional<org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer> persistentDataContainer // Paper - add persistent data container
     ) {
+        // MultiPaper start - Sync entity
+        this.id = id;
+        this.level = level;
+        // MultiPaper end - Sync entity
         this.started = started;
-        this.active = active;
+        this.active = false; // MultiPaper - Sync entity - default to false
         this.ticksActive = ticksActive;
         this.raidOmenLevel = raidOmenLevel;
         this.groupsSpawned = groupsSpawned;
@@ -173,6 +193,29 @@ public class Raid {
         this.persistentDataContainer = persistentDataContainer.orElseGet(() -> new org.bukkit.craftbukkit.persistence.CraftPersistentDataContainer(PDC_TYPE_REGISTRY)); // Paper - add persistent data container
     }
 
+    // MultiPaper start - Sync entity
+    public void setRaid(Raid raid) {
+        this.id = raid.id;
+        this.level = raid.level;
+        this.started = raid.started;
+        // default to false
+        this.active = false;
+        this.ticksActive = raid.ticksActive;
+        this.raidOmenLevel = raid.raidOmenLevel;
+        this.groupsSpawned = raid.groupsSpawned;
+        this.raidCooldownTicks = raid.raidCooldownTicks;
+        this.postRaidTicks = raid.postRaidTicks;
+        this.totalHealth = raid.totalHealth;
+        this.numGroups = raid.numGroups;
+        this.status = raid.status;
+        this.center = raid.center;
+        this.heroesOfTheVillage.clear();
+        this.heroesOfTheVillage.addAll(raid.heroesOfTheVillage);
+        this.persistentDataContainer.clear();
+        this.persistentDataContainer.putAll(raid.persistentDataContainer.getRaw());
+    }
+    // MultiPaper end - Sync entity
+
     public boolean isOver() {
         return this.isVictory() || this.isLoss();
     }
@@ -281,13 +324,14 @@ public class Raid {
         this.active = false;
         this.raidEvent.removeAllPlayers();
         this.status = Raid.RaidStatus.STOPPED;
+        puregero.multipaper.externalserverprotocol.RaidUpdatePacket.broadcastUpdate(this); // MultiPaper - Sync entity
     }
 
     public void tick(final ServerLevel level) {
         if (!this.isStopped()) {
             if (this.status == Raid.RaidStatus.ONGOING) {
                 boolean oldActive = this.active;
-                this.active = level.hasChunkAt(this.center);
+                this.active = puregero.multipaper.MultiPaper.isChunkLocal(level.getChunkIfLoaded(this.center)); // MultiPaper - Sync entity - only tick local raids
                 if (level.getDifficulty() == Difficulty.PEACEFUL) {
                     org.bukkit.craftbukkit.event.CraftEventFactory.callRaidStopEvent(level, this, org.bukkit.event.raid.RaidStopEvent.Reason.PEACE); // CraftBukkit
                     this.stop();
@@ -596,6 +640,7 @@ public class Raid {
         this.updateBossbar();
         this.setDirty(level);
         org.bukkit.craftbukkit.event.CraftEventFactory.callRaidSpawnWaveEvent(level, this, java.util.Objects.requireNonNull(this.getLeader(groupNumber)), this.groupRaiderMap.get(groupNumber)); // CraftBukkit
+        puregero.multipaper.externalserverprotocol.RaidUpdatePacket.broadcastUpdate(this); // MultiPaper - Sync entity - sync raid on new wave
     }
 
     public void joinRaid(final ServerLevel level, final int groupNumber, final Raider raider, final @Nullable BlockPos pos, final boolean exists) {
@@ -742,12 +787,13 @@ public class Raid {
         }
 
         raiders.add(raider);
-        if (updateHealth) {
+        if (updateHealth && existingCopy == null) { // MultiPaper - Sync entity - don't count health if the raider is already added
             this.totalHealth = this.totalHealth + raider.getHealth();
         }
 
         this.updateBossbar();
         this.setDirty(level);
+        puregero.multipaper.externalserverprotocol.RaidJoinPacket.broadcastJoin(raider, this.id); // MultiPaper - Sync entity
         return true;
     }
 
diff --git a/net/minecraft/world/entity/raid/Raids.java b/net/minecraft/world/entity/raid/Raids.java
index 1e509c47e779f3312dbf49428e6d4dc80f603a9c..71a839ad7c6a160c05669031fe0ecd2ae1bf3f37 100644
--- a/net/minecraft/world/entity/raid/Raids.java
+++ b/net/minecraft/world/entity/raid/Raids.java
@@ -43,6 +43,7 @@ public class Raids extends SavedData {
     );
     public static final SavedDataType<Raids> TYPE = new SavedDataType<>(RAID_FILE_ID, Raids::new, CODEC, DataFixTypes.SAVED_DATA_RAIDS);
     public final Int2ObjectMap<Raid> raidMap = new Int2ObjectOpenHashMap<>();
+    public final it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap chunkToRaidIdMap = new it.unimi.dsi.fastutil.longs.Long2IntOpenHashMap(); // MultiPaper - Sync entity
     private int nextId = 1;
     private int tick;
 
@@ -53,6 +54,7 @@ public class Raids extends SavedData {
     private Raids(final List<Raids.RaidWithId> raids, final int nextId, final int tick) {
         for (Raids.RaidWithId raid : raids) {
             this.raidMap.put(raid.id, raid.raid);
+            this.chunkToRaidIdMap.put(ChunkPos.pack(raid.raid.getCenter()), raid.id); // MultiPaper - Sync entity
             raid.raid.idOrNegativeOne = raid.id; // Paper - expose id of raids while method is kept around as deprecated for removal
         }
 
@@ -74,6 +76,16 @@ public class Raids extends SavedData {
         return OptionalInt.empty();
     }
 
+    // MultiPaper start - Sync entity
+    public java.util.Optional<Raid> getActiveRaid(ChunkPos chunkPos) {
+        return java.util.Optional
+                .of(this.chunkToRaidIdMap.getOrDefault(chunkPos.longKey(), 0))
+                .filter(id -> id != 0)
+                .map(this::get)
+                .filter(raid -> !raid.isStopped());
+    }
+    // MultiPaper end - Sync entity
+
     public void tick(final ServerLevel level) {
         this.tick++;
         // Purpur start - Raid cooldown setting
@@ -159,7 +171,11 @@ public class Raids extends SavedData {
             }if (level.purpurConfig.raidCooldownSeconds != 0) playerCooldowns.put(player.getUUID(), level.purpurConfig.raidCooldownSeconds); // Purpur - Raid cooldown setting
 
             if (!raid.isStarted() && !this.raidMap.containsValue(raid)) {
-                this.raidMap.put(this.getUniqueId(), raid);
+                // MultiPaper start - Sync entity
+                this.raidMap.put(raid.id, raid);
+                this.chunkToRaidIdMap.put(ChunkPos.pack(raid.getCenter()), raid.id);
+                puregero.multipaper.externalserverprotocol.RaidUpdatePacket.broadcastUpdate(raid);
+                // MultiPaper end - Sync entity
                 raid.idOrNegativeOne = this.nextId; // Paper - expose id of raids while method is kept around as deprecated for removal
             }
             // CraftBukkit end
@@ -171,8 +187,15 @@ public class Raids extends SavedData {
     }
 
     private Raid getOrCreateRaid(final ServerLevel level, final BlockPos pos) {
+        // MultiPaper start - Sync entity
+        // don't overlap multiple raids on the same chunk
+        java.util.Optional<Raid> optionalRaid = this.getActiveRaid(ChunkPos.containing(pos));
+        if (optionalRaid.isPresent()) {
+            return optionalRaid.get();
+        }
+        // MultiPaper end - Sync entity
         Raid raid = level.getRaidAt(pos);
-        return raid != null ? raid : new Raid(pos, level.getDifficulty());
+        return raid != null ? raid : new Raid(this.getUniqueId(), level.uuid, pos, level.getDifficulty()); // MultiPaper - Sync entity
     }
 
     public static Raids load(final CompoundTag tag) {
@@ -180,6 +203,7 @@ public class Raids extends SavedData {
     }
 
     private int getUniqueId() {
+        if (true) return new java.util.Random().nextInt(); // MultiPaper - Sync entity - 1 in 4 billion chance of two ids being the same, while not having to sync the next available id across servers
         return ++this.nextId;
     }
 
@@ -189,7 +213,7 @@ public class Raids extends SavedData {
 
         for (Raid raid : this.raidMap.values()) {
             double distance = raid.getCenter().distSqr(pos);
-            if (raid.isActive() && distance < closestDistanceSqr) {
+            if (!raid.isStopped() && distance < closestDistanceSqr) { // MultiPaper - Sync entity - use !isStopped() instead of isActive()
                 closest = raid;
                 closestDistanceSqr = distance;
             }
diff --git a/net/minecraft/world/level/block/entity/HopperBlockEntity.java b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
index fef616d7b92454a721f7a14941b62a8963080cc5..ab22ea1b08c90d9cd2ea33892192a2b50a7691c5 100644
--- a/net/minecraft/world/level/block/entity/HopperBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/HopperBlockEntity.java
@@ -479,7 +479,13 @@ public class HopperBlockEntity extends RandomizableContainerBlockEntity implemen
         BlockPos blockPos = BlockPos.containing(hopper.getLevelX(), hopper.getLevelY() + 1.0, hopper.getLevelZ());
         BlockState blockState = level.getBlockState(blockPos);
         Container container = getSourceContainer(level, hopper, blockPos, blockState);
-        if (container != null) {
+        // MultiPaper start - Sync entity
+        // don't suck from external minecarts and boats
+        if (container != null
+                && !(container instanceof Entity entity && puregero.multipaper.MultiPaperEntityInteractHandler.getOwner(entity) != null)
+                && !(container instanceof BlockEntity blockEntity && !puregero.multipaper.MultiPaper.isChunkLocal(level.getChunkIfLoaded(blockEntity.getBlockPos())))
+                && !(container instanceof net.minecraft.world.CompoundContainer compoundContainer && !puregero.multipaper.MultiPaper.isChunkLocal(level.getChunkIfLoaded(puregero.multipaper.MultiPaperContainerHandler.getPrimaryChest(compoundContainer).getBlockPos())))) {
+        // MultiPaper end - Sync entity
             Direction direction = Direction.DOWN;
             skipPullModeEventFire = skipHopperEvents; // Paper - Perf: Optimize Hoppers
 
diff --git a/net/minecraft/world/level/dimension/end/EnderDragonFight.java b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
index 26926083cbb2722bb5fbcbc81bcc44a99799ad9e..6593ab0809a20b5c30f8ffa6851a0ae7740c219d 100644
--- a/net/minecraft/world/level/dimension/end/EnderDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EnderDragonFight.java
@@ -151,7 +151,7 @@ public class EnderDragonFight extends SavedData {
         this.level = level;
         this.origin = origin;
         this.dragonEvent = new ServerBossEvent(
-            Mth.createInsecureUUID(level.getRandom()), EVENT_DISPLAY_NAME, BossEvent.BossBarColor.PINK, BossEvent.BossBarOverlay.PROGRESS
+            UUID.fromString("e3c333b8-b716-432e-8426-01d93e980678"), EVENT_DISPLAY_NAME, BossEvent.BossBarColor.PINK, BossEvent.BossBarOverlay.PROGRESS // MultiPaper - Sync entity - add specific uuid
         );
         this.dragonEvent.setPlayBossMusic(true).setCreateWorldFog(true);
         this.validPlayer = EntitySelector.ENTITY_STILL_ALIVE.and(EntitySelector.withinDistance(origin.getX(), 128 + origin.getY(), origin.getZ(), 192.0));
@@ -169,6 +169,26 @@ public class EnderDragonFight extends SavedData {
         }
     }
 
+    // MultiPaper start - Sync entity
+    public void sendDataSync() {
+        puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(new puregero.multipaper.externalserverprotocol.SetEnderDragonFightPacket(this));
+    }
+
+    public void setEnderDragonFight(EnderDragonFight enderDragonFight) {
+        this.needsStateScanning = enderDragonFight.needsStateScanning;
+        this.dragonUUID = enderDragonFight.dragonUUID;
+        this.dragonKilled = enderDragonFight.dragonKilled;
+        this.hasPreviouslyKilledDragon = enderDragonFight.hasPreviouslyKilledDragon;
+        this.respawnStage = enderDragonFight.respawnStage;
+        this.respawnTime = enderDragonFight.respawnTime;
+        this.exitPortalLocation = enderDragonFight.exitPortalLocation;
+        this.gateways.clear();
+        this.gateways.addAll(enderDragonFight.gateways);
+        this.respawnCrystals = enderDragonFight.respawnCrystals;
+        this.setDirty();
+    }
+    // MultiPaper end - Sync entity
+
     @Deprecated
     @VisibleForTesting
     public void skipArenaLoadedCheck() {
@@ -183,7 +203,12 @@ public class EnderDragonFight extends SavedData {
         }
 
         if (!this.dragonEvent.getPlayers().isEmpty()) {
-            this.level.getChunkSource().addTicketWithRadius(TicketType.DRAGON, new ChunkPos(0, 0), 9);
+            // MultiPaper start - Sync entity
+            // only request chunk ownership if we have a player there
+            if (this.dragonEvent.getPlayers().stream().anyMatch(puregero.multipaper.MultiPaper::isRealPlayer)) this.level.getChunkSource().addTicketWithRadius(TicketType.DRAGON, new ChunkPos(0, 0), 9);
+            // only tick the dragon if we own the end island
+            if (!puregero.multipaper.MultiPaper.isChunkLocal(this.level, new BlockPos(0, 0, 0))) return;
+            // MultiPaper end - Sync entity
             if (!this.isArenaLoaded()) {
                 return;
             }
@@ -259,6 +284,7 @@ public class EnderDragonFight extends SavedData {
         }
 
         this.setDirty();
+        this.sendDataSync(); // MultiPaper - Sync entity
     }
 
     private void findOrCreateDragon() {
@@ -270,6 +296,7 @@ public class EnderDragonFight extends SavedData {
             LOGGER.debug("Haven't seen our dragon, but found another one to use.");
             this.dragonUUID = entities.get(0).getUUID();
             this.setDirty();
+            this.sendDataSync(); // MultiPaper - Sync entity
         }
     }
 
@@ -293,6 +320,7 @@ public class EnderDragonFight extends SavedData {
         }
 
         this.setDirty();
+        this.sendDataSync(); // MultiPaper - Sync entity
     }
 
     private boolean hasActiveExitPortal() {
@@ -427,6 +455,7 @@ public class EnderDragonFight extends SavedData {
             this.hasPreviouslyKilledDragon = true;
             this.dragonKilled = true;
             this.setDirty();
+            this.sendDataSync(); // MultiPaper - Sync entity
         }
     }
 
@@ -454,6 +483,7 @@ public class EnderDragonFight extends SavedData {
             int z = Mth.floor(96.0 * Math.sin(2.0 * (-Math.PI + (Math.PI / 20) * gateway)));
             this.spawnNewGateway(new BlockPos(x, 75, z));
             this.setDirty();
+            this.sendDataSync(); // MultiPaper - Sync entity
             return true; // Paper - More DragonBattle API
         }
         return false; // Paper - More DragonBattle API
@@ -505,6 +535,7 @@ public class EnderDragonFight extends SavedData {
             this.dragonUUID = dragon.getUUID();
             this.resetSpikeCrystals(); // Paper - Reset ender crystals on dragon spawn
             this.setDirty();
+            this.sendDataSync(); // MultiPaper - Sync entity
         }
 
         return dragon;
@@ -546,6 +577,7 @@ public class EnderDragonFight extends SavedData {
         this.resetSpikeCrystals();
         this.spawnExitPortal(true);
         this.setDirty();
+        this.sendDataSync(); // MultiPaper - Sync entity
     }
 
     public boolean hasPreviouslyKilledDragon() {
@@ -626,6 +658,7 @@ public class EnderDragonFight extends SavedData {
             this.spawnExitPortal(false);
             this.respawnCrystals = crystals.stream().map(EntityReference::of).toList();
             this.setDirty();
+            this.sendDataSync(); // MultiPaper - Sync entity
             return true; // CraftBukkit - return value
         }
         return false; // CraftBukkit - return value
