From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koyorice <87787929+Koyorice@users.noreply.github.com>
Date: Sat, 4 Oct 2025 21:38:02 +0800
Subject: [PATCH] Sync game time


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index b4dad477fd562b26db7fd7a1764f54491bee32b8..e7987045be11f3b18e51b5a4ff98188269508c19 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -1975,7 +1975,14 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         profiler.push("timeSync");
         // Paper start - per-world time
         for (ServerLevel level : this.getAllLevels()) {
-            this.playerList.broadcastAll(new ClientboundSetTimePacket(level.getGameTime(), Map.of()), level);
+            // MultiPaper start - Sync game time
+            // don't send the time to external players
+            final ClientboundSetTimePacket packet = new ClientboundSetTimePacket(level.getGameTime(), Map.of());
+            for (final ServerPlayer player : level.players()) {
+                if (puregero.multipaper.MultiPaper.isExternalPlayer(player)) continue;
+                player.connection.send(packet);
+            }
+            // MultiPaper end - Sync game time
         }
         // Paper end - per-world time
         profiler.pop();
diff --git a/net/minecraft/server/commands/TimeCommand.java b/net/minecraft/server/commands/TimeCommand.java
index 27ec27001aad55c9e1c691e3910c08a57b3322c9..555179ff409c8c3a40f5a4c0622bcc83f1a21f6b 100644
--- a/net/minecraft/server/commands/TimeCommand.java
+++ b/net/minecraft/server/commands/TimeCommand.java
@@ -179,6 +179,7 @@ public class TimeCommand {
         if (event.callEvent()) {
             newTotalTicks = currentTotalTicks + event.getSkipAmount();
             clockManager.setTotalTicks(clock, newTotalTicks);
+            puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(new puregero.multipaper.externalserverprotocol.TimeUpdatePacket(source.getLevel(), true)); // MultiPaper - Sync game time
         } else {
             newTotalTicks = totalTicks; // just pass it through
         }
@@ -191,6 +192,7 @@ public class TimeCommand {
         org.bukkit.event.world.ClockTimeSkipEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.createTimeSkipEvent(source, time);
         if (event.callEvent()) {
             clockManager.addTicks(clock, event.getSkipAmount());
+            puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(new puregero.multipaper.externalserverprotocol.TimeUpdatePacket(source.getLevel(), true)); // MultiPaper - Sync game time
         }
         long totalTicks = clockManager.getTotalTicks(clock);
         source.sendSuccess(() -> Component.translatable("commands.time.set.absolute", clock.getRegisteredName(), totalTicks), true);
@@ -211,6 +213,7 @@ public class TimeCommand {
         final org.bukkit.event.world.ClockTimeSkipEvent event = org.bukkit.craftbukkit.event.CraftEventFactory.createTimeSkipEvent(source, targetTime.getAsLong() - currentTime);
         if (event.callEvent()) {
             clockManager.setTotalTicks(clock, currentTime + event.getSkipAmount());
+            puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(new puregero.multipaper.externalserverprotocol.TimeUpdatePacket(source.getLevel(), true)); // MultiPaper - Sync game time
         }
         // Paper end - per-world time
         source.sendSuccess(() -> Component.translatable("commands.time.set.time_marker", clock.getRegisteredName(), timeMarkerId.identifier().toString()), true);
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 1935652db020f1b1714299590802b074f6b5e804..ec7f72a05edbcabb24eae2b69eb907d3914436c9 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -845,6 +845,7 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
 
                 if (event.callEvent()) {
                     this.clockManager().setTotalTicks(defaultClock.get(), currentTime + event.getSkipAmount());
+                    puregero.multipaper.MultiPaper.broadcastPacketToExternalServers(new puregero.multipaper.externalserverprotocol.TimeUpdatePacket(this, true)); // MultiPaper - Sync game time
                 }
                 // Paper end - per-world time
             }
diff --git a/net/minecraft/world/level/saveddata/WeatherData.java b/net/minecraft/world/level/saveddata/WeatherData.java
index 93adfc0e79dc52c602f5b9f258b4ee1d8b852eb8..54f088928b993e487fbf425cc4b40eb2eaae1ab2 100644
--- a/net/minecraft/world/level/saveddata/WeatherData.java
+++ b/net/minecraft/world/level/saveddata/WeatherData.java
@@ -71,6 +71,7 @@ public final class WeatherData extends SavedData {
         // CraftBukkit end
         this.thundering = thundering;
         this.setDirty();
+        if (this.level != null) puregero.multipaper.MultiPaperRainHandler.onWeatherChange(this, this.level.uuid); // MultiPaper - Sync game time
     }
 
     public int getThunderTime() {
@@ -109,6 +110,7 @@ public final class WeatherData extends SavedData {
         // CraftBukkit end
         this.raining = raining;
         this.setDirty();
+        if (this.level != null) puregero.multipaper.MultiPaperRainHandler.onWeatherChange(this, this.level.uuid); // MultiPaper - Sync game time
     }
 
     public int getRainTime() {
