--- a/purpur-server/build.gradle.kts
+++ b/purpur-server/build.gradle.kts
@@ -33,9 +_,33 @@
             outputDir = rootDirectory.dir("paper-server")
         }
     }
-    activeFork = purpur
-    // Purpur end - Rebrand
+    // activeFork = purpur // MultiFolia
 
-    // MultiPaper start - Rebrand to MultiPaper
-    val multipaper = forks.register("multipaper") {
+    // MultiFolia start - server integration fork
+    val multifolia = forks.register("multifolia") {
         forks = purpur
         upstream.patchRepo("paperServer") {
             upstreamRepo = purpur.patchedRepo("paperServer")
-            patchesDir = rootDirectory.dir("multipaper-server/paper-patches")
+            patchesDir = rootDirectory.dir("multipaper-server/paper-patches")
             outputDir = rootDirectory.dir("paper-server")
         }
         upstream.patchDir("purpurServer") {
             upstreamPath = "purpur-server"
             excludes = setOf(
@@ -48,8 +_,8 @@
             outputDir = rootDirectory.dir("purpur-server")
         }
     }
-    activeFork = multipaper
-    // MultiPaper end - Rebrand to MultiPaper
+    activeFork = multifolia
+    // MultiFolia end - server integration fork
@@ -105,10 +_,14 @@
     main {
         java { srcDir("../paper-server/src/main/java") }
         resources { srcDir("../paper-server/src/main/resources") }
+        java { srcDir("../purpur-server/src/main/java") }
+        resources { srcDir("../purpur-server/src/main/resources") }
     }
     test {
         java { srcDir("../paper-server/src/test/java") }
         resources { srcDir("../paper-server/src/test/resources") }
+        java { srcDir("../purpur-server/src/test/java") }
+        resources { srcDir("../purpur-server/src/test/resources") }
     }
 }
 val log4jPlugins = sourceSets.create("log4jPlugins") {
@@ -137,14 +_,14 @@
 }
 
 dependencies {
-    implementation(project(":purpur-api")) // Purpur
+    // MultiFolia start
+    implementation(project(":multipaper-api"))
+    implementation(project(":multipaper-mastermessagingprotocol"))
+    implementation("io.methvin:directory-watcher:0.19.1")
+    implementation("com.github.luben:zstd-jni:1.5.7-4")
+    // MultiFolia end
     implementation("ca.spottedleaf:leafpile:1.0.0")
     implementation("org.jline:jline-terminal-ffm:3.27.1") // use ffm on java 22+
     implementation("org.jline:jline-terminal-jni:3.27.1") // fall back to jni on java 21
@@ -203,14 +_,14 @@
         val gitBranch = git.exec(providers, "rev-parse", "--abbrev-ref", "HEAD").get().trim()
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
-            "Implementation-Title" to "MultiPaper", // Purpur // MultiPaper
+            "Implementation-Title" to "MultiFolia",
             "Implementation-Version" to implementationVersion,
             "Implementation-Vendor" to date,
-            "Specification-Title" to "MultiPaper", // Purpur // MultiPaper
+            "Specification-Title" to "MultiFolia",
             "Specification-Version" to project.version,
-            "Specification-Vendor" to "MultiPaper", // Purpur // MultiPaper
-            "Brand-Id" to "multipaper:multipaper", // Purpur // MultiPaper
-            "Brand-Name" to "MultiPaper", // Purpur // MultiPaper
+            "Specification-Vendor" to "MultiFolia",
+            "Brand-Id" to "multifolia:multifolia",
+            "Brand-Name" to "MultiFolia",
             "Build-Number" to (build ?: ""),
             "Build-Time" to buildTime.toString(),
             "Git-Branch" to gitBranch,
