From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Koyorice <87787929+Koyorice@users.noreply.github.com>
Date: Thu, 9 Oct 2025 23:27:30 +0800
Subject: [PATCH] Add MultiPaper API


diff --git a/src/main/java/com/destroystokyo/paper/Title.java b/src/main/java/com/destroystokyo/paper/Title.java
index 93b2f82bbb1276bf21ac0aa7e88a27c5a835a51c..94131c0f76be18401d02735b0df35c5f21480400 100644
--- a/src/main/java/com/destroystokyo/paper/Title.java
+++ b/src/main/java/com/destroystokyo/paper/Title.java
@@ -268,7 +268,7 @@ public final class Title {
      * Sends the title directly to all online players
      */
     public void broadcast() {
-        send(Bukkit.getOnlinePlayers());
+        send(Bukkit.getAllOnlinePlayers()); // MultiPaper - Add MultiPaper API
     }
 
     @NotNull
diff --git a/src/main/java/com/destroystokyo/paper/event/server/PaperServerListPingEvent.java b/src/main/java/com/destroystokyo/paper/event/server/PaperServerListPingEvent.java
index 15b12abe17dadf0fc9eeee36de4b2ff2b1b52cbc..010f576214ecff744918c1f5517b9be45406d5e4 100644
--- a/src/main/java/com/destroystokyo/paper/event/server/PaperServerListPingEvent.java
+++ b/src/main/java/com/destroystokyo/paper/event/server/PaperServerListPingEvent.java
@@ -293,7 +293,7 @@ public class PaperServerListPingEvent extends ServerListPingEvent implements Can
     }
 
     protected @NotNull Object @NotNull [] getOnlinePlayers() {
-        return Bukkit.getOnlinePlayers().toArray();
+        return Bukkit.getAllOnlinePlayers().toArray(); // MultiPaper - Add MultiPaper API
     }
 
     @NotNull
diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 2c39beef3747d0dfd15a6e08269382e7b5a55d36..45ce5693b0577d04e199af6bbc3ae4faf70a99e3 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -173,8 +173,11 @@ public final class Bukkit {
     }
     // Paper end
 
+    // MultiPaper start - Add MultiPaper API
     /**
-     * Gets a view of all currently logged in players. This {@linkplain
+     * Gets a view of all currently logged in players on your local
+     * MultiPaper instance. This method may return the logged in players
+     * across all instances when tab-completing a command. This {@linkplain
      * Collections#unmodifiableCollection(Collection) view} is a reused
      * object, making some operations like {@link Collection#size()}
      * zero-allocation.
@@ -205,6 +208,73 @@ public final class Bukkit {
         return server.getOnlinePlayers();
     }
 
+    /**
+     * Gets a view of all currently logged in players on your local
+     * MultiPaper instance. This {@linkplain
+     * Collections#unmodifiableCollection(Collection) view} is a reused
+     * object, making some operations like {@link Collection#size()}
+     * zero-allocation.
+     * <p>
+     * The collection is a view backed by the internal representation, such
+     * that, changes to the internal state of the server will be reflected
+     * immediately. However, the reuse of the returned collection (identity)
+     * is not strictly guaranteed for future or all implementations. Casting
+     * the collection, or relying on interface implementations (like {@link
+     * Serializable} or {@link List}), is deprecated.
+     * <p>
+     * Iteration behavior is undefined outside of self-contained main-thread
+     * uses. Normal and immediate iterator use without consequences that
+     * affect the collection are fully supported. The effects following
+     * (non-exhaustive) {@link Entity#teleport(Location) teleportation},
+     * {@link Player#setHealth(double) death}, and {@link Player#kick(
+     * Component) kicking} are undefined. Any use of this collection from
+     * asynchronous threads is unsafe.
+     * <p>
+     * For safe consequential iteration or mimicking the old array behavior,
+     * using {@link Collection#toArray(Object[])} is recommended. For making
+     * snapshots, {@link ImmutableList#copyOf(Collection)} is recommended.
+     *
+     * @return a view of currently online players.
+     */
+    @NotNull
+    public static Collection<? extends Player> getLocalOnlinePlayers() {
+        return server.getLocalOnlinePlayers();
+    }
+
+    /**
+     * Gets a view of all currently logged in players across all
+     * MultiPaper instances. This {@linkplain
+     * Collections#unmodifiableCollection(Collection) view} is a reused
+     * object, making some operations like {@link Collection#size()}
+     * zero-allocation.
+     * <p>
+     * The collection is a view backed by the internal representation, such
+     * that, changes to the internal state of the server will be reflected
+     * immediately. However, the reuse of the returned collection (identity)
+     * is not strictly guaranteed for future or all implementations. Casting
+     * the collection, or relying on interface implementations (like {@link
+     * Serializable} or {@link List}), is deprecated.
+     * <p>
+     * Iteration behavior is undefined outside of self-contained main-thread
+     * uses. Normal and immediate iterator use without consequences that
+     * affect the collection are fully supported. The effects following
+     * (non-exhaustive) {@link Entity#teleport(Location) teleportation},
+     * {@link Player#setHealth(double) death}, and {@link Player#kick(
+     * Component) kicking} are undefined. Any use of this collection from
+     * asynchronous threads is unsafe.
+     * <p>
+     * For safe consequential iteration or mimicking the old array behavior,
+     * using {@link Collection#toArray(Object[])} is recommended. For making
+     * snapshots, {@link ImmutableList#copyOf(Collection)} is recommended.
+     *
+     * @return a view of currently online players.
+     */
+    @NotNull
+    public static Collection<? extends Player> getAllOnlinePlayers() {
+        return server.getAllOnlinePlayers();
+    }
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Get the maximum amount of players which can login to this server.
      *
@@ -789,6 +859,39 @@ public final class Bukkit {
         return server.getServicesManager();
     }
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Gets the multipaper key-value data storage. Accessing this data is
+     * asynchronous. This storage medium is hosted on the Master instance.
+     *
+     * @return the multipaper data storage
+     */
+    @NotNull
+    public static puregero.multipaper.MultiPaperDataStorage getMultiPaperDataStorage() {
+        return server.getMultiPaperDataStorage();
+    }
+
+    /**
+     * Gets a multipaper notification manager.
+     *
+     * @return a multipaper notification manager
+     */
+    @NotNull
+    public static puregero.multipaper.MultiPaperNotificationManager getMultiPaperNotificationManager() {
+        return server.getMultiPaperNotificationManager();
+    }
+
+    /**
+     * Get the MultiPaper name of this server.
+     *
+     * @return the MultiPaper name of this server
+     */
+    @NotNull
+    public static String getLocalServerName() {
+        return server.getLocalServerName();
+    }
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Gets a list of all worlds on this server.
      *
diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index d873b78f6935a23ff8f6092b80968267b182cc0f..d79104ec6d1544d46ebb9f41ec7421dcd9288960 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -222,6 +222,24 @@ public interface Chunk extends PersistentDataHolder {
      */
     boolean isSlimeChunk();
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    boolean isExternalChunk();
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    boolean isLocalChunk();
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Gets whether the chunk at the specified chunk coordinates is force
      * loaded.
diff --git a/src/main/java/org/bukkit/Location.java b/src/main/java/org/bukkit/Location.java
index 1a89e738b7923553d1080ecd497ad60d34dd0f37..972111bd6d22cc202b9497197c17601798856316 100644
--- a/src/main/java/org/bukkit/Location.java
+++ b/src/main/java/org/bukkit/Location.java
@@ -344,6 +344,28 @@ public class Location implements Cloneable, ConfigurationSerializable, io.paperm
         return this;
     }
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isChunkExternal() {
+        return this.getWorld().isChunkExternal(locToBlock(this.x) >> 4, locToBlock(this.z) >> 4);
+    }
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isChunkLocal() {
+        return this.getWorld().isChunkLocal(locToBlock(this.x) >> 4, locToBlock(this.z) >> 4);
+    }
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Adds the location by another.
      *
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index d277417d2bd66b84f293e9b3858b5096a3a86fe4..fb22e801b6a48f98ab4b24c8f720ab7653d3b865 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -135,8 +135,11 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     String getMinecraftVersion();
     // Paper end
 
+    // MultiPaper start - Add MultiPaper API
     /**
-     * Gets a view of all currently logged in players. This {@linkplain
+     * Gets a view of all currently logged in players on your local
+     * MultiPaper instance. This method may return the logged in players
+     * across all instances when tab-completing a command. This {@linkplain
      * Collections#unmodifiableCollection(Collection) view} is a reused
      * object, making some operations like {@link Collection#size()}
      * zero-allocation.
@@ -165,6 +168,69 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public Collection<? extends Player> getOnlinePlayers();
 
+    /**
+     * Gets a view of all currently logged in players on your local
+     * MultiPaper instance. This {@linkplain
+     * Collections#unmodifiableCollection(Collection) view} is a reused
+     * object, making some operations like {@link Collection#size()}
+     * zero-allocation.
+     * <p>
+     * The collection is a view backed by the internal representation, such
+     * that, changes to the internal state of the server will be reflected
+     * immediately. However, the reuse of the returned collection (identity)
+     * is not strictly guaranteed for future or all implementations. Casting
+     * the collection, or relying on interface implementations (like {@link
+     * Serializable} or {@link List}), is deprecated.
+     * <p>
+     * Iteration behavior is undefined outside of self-contained main-thread
+     * uses. Normal and immediate iterator use without consequences that
+     * affect the collection are fully supported. The effects following
+     * (non-exhaustive) {@link Entity#teleport(Location) teleportation},
+     * {@link Player#setHealth(double) death}, and {@link Player#kick(
+     * Component) kicking} are undefined. Any use of this collection from
+     * asynchronous threads is unsafe.
+     * <p>
+     * For safe consequential iteration or mimicking the old array behavior,
+     * using {@link Collection#toArray(Object[])} is recommended. For making
+     * snapshots, {@link ImmutableList#copyOf(Collection)} is recommended.
+     *
+     * @return a view of currently online players.
+     */
+    @NotNull
+    Collection<? extends Player> getLocalOnlinePlayers();
+
+    /**
+     * Gets a view of all currently logged in players across all
+     * MultiPaper instances. This {@linkplain
+     * Collections#unmodifiableCollection(Collection) view} is a reused
+     * object, making some operations like {@link Collection#size()}
+     * zero-allocation.
+     * <p>
+     * The collection is a view backed by the internal representation, such
+     * that, changes to the internal state of the server will be reflected
+     * immediately. However, the reuse of the returned collection (identity)
+     * is not strictly guaranteed for future or all implementations. Casting
+     * the collection, or relying on interface implementations (like {@link
+     * Serializable} or {@link List}), is deprecated.
+     * <p>
+     * Iteration behavior is undefined outside of self-contained main-thread
+     * uses. Normal and immediate iterator use without consequences that
+     * affect the collection are fully supported. The effects following
+     * (non-exhaustive) {@link Entity#teleport(Location) teleportation},
+     * {@link Player#setHealth(double) death}, and {@link Player#kick(
+     * Component) kicking} are undefined. Any use of this collection from
+     * asynchronous threads is unsafe.
+     * <p>
+     * For safe consequential iteration or mimicking the old array behavior,
+     * using {@link Collection#toArray(Object[])} is recommended. For making
+     * snapshots, {@link ImmutableList#copyOf(Collection)} is recommended.
+     *
+     * @return a view of currently online players.
+     */
+    @NotNull
+    Collection<? extends Player> getAllOnlinePlayers();
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Get the maximum amount of players which can login to this server.
      *
@@ -716,6 +782,33 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
     @NotNull
     public ServicesManager getServicesManager();
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Gets the multipaper key-value data storage. Accessing this data is
+     * asynchronous. This storage medium is hosted on the Master instance.
+     *
+     * @return the multipaper data storage
+     */
+    @NotNull
+    puregero.multipaper.MultiPaperDataStorage getMultiPaperDataStorage();
+
+    /**
+     * Gets a multipaper notification manager.
+     *
+     * @return a multipaper notification manager
+     */
+    @NotNull
+    puregero.multipaper.MultiPaperNotificationManager getMultiPaperNotificationManager();
+
+    /**
+     * Get the MultiPaper name of this server.
+     *
+     * @return the MultiPaper name of this server
+     */
+    @NotNull
+    String getLocalServerName();
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Gets a list of all worlds on this server.
      *
diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index c5d6067760403d85d38f03b51ff4aa476eff26f3..bd76816de64f8e9a4d07b7c164e01bce6fc82e67 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -409,6 +409,24 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
     @Deprecated(since = "1.14")
     public boolean isChunkInUse(int x, int z);
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Returns whether the chunk is running on an external server or not.
+     *
+     * @return True if the chunk is an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    public boolean isChunkExternal(int x, int z);
+
+    /**
+     * Returns whether the chunk is running on this server or not.
+     *
+     * @return True if the chunk is a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    public boolean isChunkLocal(int x, int z);
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Loads the {@link Chunk} at the specified coordinates.
      * <p>
diff --git a/src/main/java/org/bukkit/block/Block.java b/src/main/java/org/bukkit/block/Block.java
index 5c75b3d049f22573c4bc85ed06fc7f1efa31b361..3045fedaa4c4dd170fd0212e6e1177266baa54b4 100644
--- a/src/main/java/org/bukkit/block/Block.java
+++ b/src/main/java/org/bukkit/block/Block.java
@@ -449,6 +449,26 @@ public interface Block extends Metadatable, Translatable, net.kyori.adventure.tr
         return this.getBlockPower(BlockFace.SELF);
     }
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Returns whether the block is in a chunk that's running on an external
+     * server or not.
+     *
+     * @return True if the block is in an external chunk, or false if the chunk
+     * is running on this server or if it's unloaded.
+     */
+    boolean isInExternalChunk();
+
+    /**
+     * Returns whether the block is in a chunk that's running on this server
+     * or not.
+     *
+     * @return True if the block is in a local chunk, or false if the chunk
+     * is running on an external server or if it's unloaded.
+     */
+    boolean isInLocalChunk();
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Checks if this block is empty.
      * <p>
diff --git a/src/main/java/org/bukkit/command/Command.java b/src/main/java/org/bukkit/command/Command.java
index 3171a20c8fcff10a8da3dd822c8bba3f58f59e02..908d612de4e02cf7eeaf168d25ca27fd96760b03 100644
--- a/src/main/java/org/bukkit/command/Command.java
+++ b/src/main/java/org/bukkit/command/Command.java
@@ -115,7 +115,7 @@ public abstract class Command {
         Player senderPlayer = sender instanceof Player ? (Player) sender : null;
 
         ArrayList<String> matchedPlayers = new ArrayList<String>();
-        for (Player player : sender.getServer().getOnlinePlayers()) {
+        for (Player player : sender.getServer().getAllOnlinePlayers()) { // MultiPaper - Add MultiPaper API
             String name = player.getName();
             if ((senderPlayer == null || senderPlayer.canSee(player)) && StringUtil.startsWithIgnoreCase(name, lastWord)) {
                 matchedPlayers.add(name);
diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index fcfde5d11a934c31bbf91d40f5a382d4ab329a8e..19507e29cd3587764d308c970da812f067a8755c 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -348,6 +348,24 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     public int getEntityId();
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Returns whether the entity is in a chunk that's running on an external
+     * server or not.
+     *
+     * @return True if the entity is in an external chunk.
+     */
+    boolean isInExternalChunk();
+
+    /**
+     * Returns whether the entity is in a chunk that's running on this server
+     * or not.
+     *
+     * @return True if the entity is in a local chunk.
+     */
+    boolean isInLocalChunk();
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Returns the entity's current fire ticks (ticks before the entity stops
      * being on fire).
@@ -1399,4 +1417,22 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Persistent
      */
     void setImmuneToFire(@Nullable Boolean fireImmune);
     // Purpur end - Fire Immunity API
+    // MultiPaper start - Add MultiPaper API
+    // add fake entity api
+    /**
+     * Sets the entity as a fake entity
+     * <p>
+     * once set external players can't interact or see this entity.
+     * <p>
+     * Warning: once set, this can't be changed.
+     *
+     * @throws IllegalStateException when online player is passed
+     */
+    void setFake();
+
+    /**
+     * returns whatever the entity is a fake entity
+     */
+    boolean isFake();
+    // MultiPaper end - Add MultiPaper API
 }
diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 81559e992bb344b7a71933447e40bd0ecc7a0c96..172884770d256a3b01115734ec2dcf7045ce2868 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -644,6 +644,82 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public Input getCurrentInput();
 
+    // MultiPaper start - Add MultiPaper API
+    /**
+     * Returns whether the player is on an external server or not.
+     *
+     * @return True if the player is on an external server.
+     */
+    boolean isExternalPlayer();
+
+    /**
+     * Returns whether the player is on this server or not.
+     *
+     * @return True if the player is on this server.
+     */
+    boolean isLocalPlayer();
+
+    /**
+     * Returns cross-server data stored under the specified key. Note
+     * that all plugins share the same set of keys. This data is
+     * non-persistent, it will be lost when the player disconnects.
+     *
+     * @param key The key the data is stored under.
+     * @return The data stored under the key, or null if the key isn't set.
+     */
+    @Nullable
+    String getData(String key);
+
+    /**
+     * Store cross-server data under the specified key. Note that all plugins
+     * share the same set of keys. This data is non-persistent, it will be
+     * lost when the player disconnects.
+     *
+     * @param key   The key to store the data under.
+     * @param value The data to store under the key.
+     */
+    void setData(String key, String value);
+
+    /**
+     * Returns cross-server data stored under the specified key. Note
+     * that all plugins share the same set of keys. This data is persistent,
+     * it will be saved even if the player disconnects. This persistent data is
+     * saved onto the player's .dat file.
+     *
+     * @param key The key the data is stored under.
+     * @return The data stored under the key, or null if the key isn't set.
+     */
+    @Nullable
+    String getPersistentData(String key);
+
+    /**
+     * Store cross-server data under the specified key. Note that all plugins
+     * share the same set of keys. This data is persistent, it will be saved
+     * even if the player disconnects. This persistent data is saved onto the
+     * player's .dat file.
+     *
+     * @param key   The key to store the data under.
+     * @param value The data to store under the key.
+     */
+    void setPersistentData(String key, String value);
+
+    /**
+     * Says a message (or runs a command) on another server excluding this one.
+     *
+     * @param msg chat message to say
+     */
+    void chatOnOtherServers(String msg);
+
+    /**
+     * Get the bungeecord name of the server that this player is on.
+     *
+     * @return The bungeecord name of the server the player is on for external
+     * players, or null for local players.
+     */
+    @Nullable
+    String getExternalServerName();
+    // MultiPaper end - Add MultiPaper API
+
     /**
      * Play a note for the player at a location. <br>
      * This <i>will</i> work with cake.
diff --git a/src/main/java/org/bukkit/event/player/PlayerChatEvent.java b/src/main/java/org/bukkit/event/player/PlayerChatEvent.java
index 5c5849dbc33bce83b81aab9aee2802b1fb1cdfcf..8628e8203e6a88bc79720d3bb8e2e80e6f14f940 100644
--- a/src/main/java/org/bukkit/event/player/PlayerChatEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerChatEvent.java
@@ -32,7 +32,7 @@ public class PlayerChatEvent extends PlayerEvent implements Cancellable {
         super(player);
         this.message = message;
         this.format = "<%1$s> %2$s";
-        this.recipients = new HashSet<>(player.getServer().getOnlinePlayers());
+        this.recipients = new HashSet<>(player.getServer().getAllOnlinePlayers()); // MultiPaper - Add MultiPaper API
     }
 
     @ApiStatus.Internal
diff --git a/src/main/java/org/bukkit/event/player/PlayerCommandPreprocessEvent.java b/src/main/java/org/bukkit/event/player/PlayerCommandPreprocessEvent.java
index d9a6ebfb636c5eeeae8b39b29d0f4be22b4f63b7..6d03419b53a21d02fe1e9d5081fbab93717b0e2d 100644
--- a/src/main/java/org/bukkit/event/player/PlayerCommandPreprocessEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerCommandPreprocessEvent.java
@@ -58,7 +58,7 @@ public class PlayerCommandPreprocessEvent extends PlayerEvent implements Cancell
     @ApiStatus.Internal
     public PlayerCommandPreprocessEvent(@NotNull final Player player, @NotNull final String message) {
         super(player);
-        this.recipients = new HashSet<>(player.getServer().getOnlinePlayers());
+        this.recipients = new HashSet<>(player.getServer().getAllOnlinePlayers()); // MultiPaper - Add MultiPaper API
         this.message = message;
     }
 
